{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4caadfc3-8e2f-461e-a637-aee4db5d7441",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 352734 × 45453\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'harm_study', 'harm_healthy.tissue', 'harm_tumor.site', 'harm_sample.type', 'harm_condition', 'harm_tumor.type', 'harm_cd45pos', 'harm_healthy.pat', 'percent.mt', 'ratio_nCount_nFeature', 'batch', 'X_scvi_batch', 'X_scvi_labels', 'X_scvi_local_l_mean', 'X_scvi_local_l_var', 'leiden_0.2', 'leiden_0.4', 'leiden_0.6', 'leiden_0.8', 'leiden_1', 'leiden_1.2', 'leiden_1.4', 'author_first_cell_type', 'author_cell_type', 'cnv_score', 'organism_ontology_term_id', 'donor_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'suspension_type', 'assay_ontology_term_id', 'cell_id', 'dup', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
      "    var: 'features', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type'\n",
      "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
      "    obsm: 'X_umap'\n"
     ]
    }
   ],
   "source": [
    "import anndata\n",
    "\n",
    "# Load the .h5ad file\n",
    "adata = anndata.read_h5ad(\"897e76b2-59f6-482c-827d-37cc62fa4f50.h5ad\")\n",
    "\n",
    "# Check basic details about the dataset\n",
    "print(adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9cd2ab19-eabf-462c-a1f0-23825e640f3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/preprocessing/_highly_variable_genes.py:74: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in divide\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:458: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:435: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:437: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:440: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:450: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Included marker genes: 300 / 311\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mygene import MyGeneInfo\n",
    "\n",
    "# -----------------------------\n",
    "# Step 0: Remove mitochondrial genes\n",
    "# -----------------------------\n",
    "mt_mask = adata.var[\"feature_name\"].str.upper().str.startswith(\"MT-\")\n",
    "adata = adata[:, ~mt_mask].copy()\n",
    "\n",
    "# -----------------------------\n",
    "# Step 1: Filter samples\n",
    "# -----------------------------\n",
    "adata = adata[adata.obs[\"disease\"] != \"normal\"].copy()\n",
    "adata = adata[adata.obs[\"cell_type\"] != \"megakaryocyte\"].copy()\n",
    "\n",
    "# -----------------------------\n",
    "# Step 2: Define group and filter rare groups\n",
    "# -----------------------------\n",
    "adata.obs[\"group\"] = adata.obs[\"cell_type\"].astype(str) + \"_\" + adata.obs[\"disease\"].astype(str)\n",
    "group_counts = adata.obs[\"group\"].value_counts()\n",
    "valid_groups = group_counts[group_counts >= 2].index\n",
    "adata = adata[adata.obs[\"group\"].isin(valid_groups)].copy()\n",
    "\n",
    "# -----------------------------\n",
    "# Step 3: Normalize and log1p\n",
    "# -----------------------------\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# -----------------------------\n",
    "# Step 4: HVG selection\n",
    "# -----------------------------\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor=\"seurat_v3\")\n",
    "\n",
    "# -----------------------------\n",
    "# Step 5: Differential expression on HVGs only\n",
    "# -----------------------------\n",
    "adata_hvg = adata[:, adata.var[\"highly_variable\"]].copy()\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata_hvg,\n",
    "    groupby=\"group\",\n",
    "    method=\"wilcoxon\",\n",
    "    n_genes=50\n",
    ")\n",
    "\n",
    "# -----------------------------\n",
    "# Step 6: Collect marker genes (Ensembl IDs)\n",
    "# -----------------------------\n",
    "marker_genes_by_group = {}\n",
    "for group in adata_hvg.obs[\"group\"].unique():\n",
    "    marker_genes_by_group[group] = list(adata_hvg.uns[\"rank_genes_groups\"][\"names\"][group][:10])\n",
    "\n",
    "# -----------------------------\n",
    "# Step 7: Map Ensembl to gene symbols\n",
    "# -----------------------------\n",
    "mg = MyGeneInfo()\n",
    "unmapped = set(g for genes in marker_genes_by_group.values() for g in genes if g.startswith(\"ENSG\"))\n",
    "query_result = mg.querymany(list(unmapped), scopes=\"ensembl.gene\", fields=\"symbol\", species=\"human\")\n",
    "external_map = {item[\"query\"]: item.get(\"symbol\", item[\"query\"]) for item in query_result}\n",
    "ensembl_to_symbol = adata.var[\"feature_name\"].to_dict()\n",
    "ensembl_to_symbol.update(external_map)\n",
    "\n",
    "# -----------------------------\n",
    "# Step 8: Rebuild gene symbol marker dict\n",
    "# -----------------------------\n",
    "converted_marker_genes_by_group = {}\n",
    "for group, gene_list in marker_genes_by_group.items():\n",
    "    converted = [ensembl_to_symbol.get(g, g) for g in gene_list]\n",
    "    converted_marker_genes_by_group[group] = converted\n",
    "\n",
    "# -----------------------------\n",
    "# Step 9: Merge HVGs and marker genes for downstream\n",
    "# -----------------------------\n",
    "# 1. Flatten marker gene list\n",
    "all_marker_genes = set(g for genes in converted_marker_genes_by_group.values() for g in genes)\n",
    "\n",
    "# 2. Make sure they are in the dataset\n",
    "gene_symbols = adata.var[\"feature_name\"]\n",
    "marker_genes_in_data = [g for g in all_marker_genes if g in gene_symbols.values]\n",
    "\n",
    "# 3. Create mask for HVGs + marker genes\n",
    "hvg_mask = adata.var[\"highly_variable\"]\n",
    "additional_genes_mask = gene_symbols.isin(marker_genes_in_data)\n",
    "final_mask = hvg_mask | additional_genes_mask\n",
    "\n",
    "# 4. Subset adata\n",
    "adata = adata[:, final_mask].copy()\n",
    "\n",
    "# -----------------------------\n",
    "# Optional: Report\n",
    "# -----------------------------\n",
    "included = [g for g in all_marker_genes if g in adata.var[\"feature_name\"].values]\n",
    "print(f\" Included marker genes: {len(included)} / {len(all_marker_genes)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f2bedff6-2ae3-4b0d-b38e-585752e84ee5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final existing marker genes:\n",
      "['GLUL', 'MIF', 'ENG', 'ESAM', 'TPSAB1', 'CLN8', 'ORAI2', 'CXCR4', 'CD52', 'PPP1R14A', 'IKZF3', 'SRGN', 'TM4SF4', 'SPARCL1', 'RAMP2', 'PECAM1', 'NAPSA', 'VWA5A', 'HBB', 'IGKC', 'VWF', 'RBP1', 'IGHV3-23', 'TNFAIP3', 'ORC4', 'MALAT1', 'ANKRD20A9P', 'ACTG1', 'RPL21', 'KRT7', 'HLA-DQA1', 'B2M', 'ZNF106', 'S100P', 'CD68', 'TCF4', 'NKG7', 'COL3A1', 'PTPRC', 'MS4A1', 'EEF1A1', 'SEC61G', 'MZB1', 'RAMP3', 'CORO1A', 'HBA1', 'SPARC', 'ALOX5AP', 'IGLC3', 'CD2', 'SHISA9', 'C15orf48', 'CRIP2', 'LGALS1', 'RAB38', 'HSP90AA1', 'ACTB', 'FCER1G', 'KRT19', 'ITM2C', 'C1S', 'MLANA', 'SCGB3A2', 'VIM', 'TMSB4X', 'H3-3A', 'ANXA5', 'FTL', 'FABP4', 'MUCL1', 'LYZ', 'SFTPA2', 'GZMB', 'MPEG1', 'PLAUR', 'FKBP1A', 'C12orf75', 'FCGR3B', 'C1R', 'ELOB', 'PTGDS', 'HOPX', 'TPSD1', 'TEX41', 'CTSS', 'MGST1', 'JCHAIN', 'PPDPF', 'EPCAM', 'HBA2', 'TCEA1', 'TFF3', 'BCL2A1', 'IRF4', 'VTN', 'HLA-C', 'CD3D', 'CST7', 'CSF3R', 'WFDC2', 'SEC61B', 'RGS13', 'KRT18', 'LUM', 'BGN', 'S100A9', 'HSPG2', 'RGS1', 'SFTPA1', 'RGS5', 'CD37', 'LILRA4', 'CFLAR', 'C19orf33', 'PLVAP', 'FXYD2', 'DCN', 'CPA3', 'IRF7', 'SAT1', 'AGR2', 'HERPUD1', 'SDCBP', 'PLAC9', 'SELENOW', 'PLAC8', 'EPAS1', 'RPL8', 'PHGR1', 'CAVIN2', 'RPS27', 'SERPINA1', 'COX6C', 'IGKV3-20', 'ALDOA', 'SNHG7', 'TSPAN13', 'EFHD1', 'GNG11', 'COL1A2', 'UQCRQ', 'G0S2', 'NAMPT', 'AZGP1', 'MNDA', 'TNFAIP8L1', 'ATP5MF', 'YWHAE', 'HLA-DRB1', 'ADGRL4', 'RARRES2', 'COL6A1', 'TRBC1', 'COL1A1', 'LITAF', 'PMEL', 'MS4A2', 'TTC19', 'MBOAT1', 'RPLP2', 'IGHG3', 'SELENOM', 'SPI1', 'RPS17', 'SERPINF1', 'COL6A3', 'HLA-DPB1', 'ASTN2', 'S100A8', 'HLA-DRA', 'CXCL8', 'GPR183', 'CLDN10', 'TPM2', 'HLA-DRB5', 'IL1RL1', 'UGDH-AS1', 'TMEM212', 'CD3E', 'REST', 'CST3', 'PERP', 'LTB', 'TAGLN', 'AIF1', 'CCDC50', 'RAC2', 'LTC4S', 'TYROBP', 'GZMA', 'CD79A', 'RPS29', 'TAPBP', 'SLCO4A1-AS1', 'S100A6', 'PELATON', 'RBMS2', 'HLA-DPA1', 'ALB', 'RPL13AP5', 'NCF1', 'S100A1', 'RPL27A', 'PPP1R14B', 'MXD1', 'SFTA2', 'LAPTM5', 'ID2', 'CHP1', 'IGKV4-1', 'GABARAP', 'TXN', 'IGLC2', 'UGCG', 'MS4A7', 'IGHGP', 'GATA2', 'RPL30', 'MYL9', 'HPGDS', 'RPS10', 'MGP', 'CD79B', 'IGFBP7', 'GATD1', 'MITF', 'SPP1', 'IGLV1-51', 'CD74', 'ANXA4', 'PDZK1IP1', 'FTH1', 'SLC18A2', 'RPL7', 'AQP1', 'LGALS4', 'TIMP1', 'CD7', 'METTL21A', 'TRBC2', 'TPSB2', 'CLDN5', 'SAMSN1', 'FPR1', 'CITED1', 'TM4SF1', 'TACSTD2', 'IFITM1', 'PLEK', 'KCNQ1OT1', 'RIPOR2', 'IFITM3', 'EGFL7', 'RPL39', 'HSP90AB1', 'CLEC14A', 'TOR1AIP2', 'FXYD3', 'BTG1', 'TYRP1', 'PLD4', 'IGHG4', 'ATP1B1', 'COL6A2', 'IL32', 'KIT', 'CD34', 'GAPDH', 'HDC', 'HLA-B', 'CALD1', 'GPX1', 'CD24', 'LST1', 'IRF8', 'CYBA', 'SCGB2A2', 'CREM', 'GSTP1', 'RGS2', 'CTSG', 'RPL17', 'HYAL2', 'ST3GAL4', 'EEF1G', 'SERPING1', 'FRZB', 'COX6A1', 'AREG', 'KRT8', 'SFTPB', 'SPINT2', 'LINC00623', 'IGHG1', 'TIMP3', 'HPGD', 'AEBP1', 'TSPAN8', 'MDK', 'SSR4']\n"
     ]
    }
   ],
   "source": [
    "print(\"Final existing marker genes:\")\n",
    "print(included)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3b0fb468-546e-4f96-8dd2-e5ee1d5f181d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "endothelial cell_lung cancer: ['VWF', 'EPAS1', 'PECAM1', 'CLDN5', 'CLEC14A', 'SPARCL1', 'RAMP2', 'HYAL2', 'AQP1', 'TM4SF1']\n",
      "mononuclear phagocyte_lung cancer: ['TYROBP', 'LYZ', 'HLA-DRA', 'FCER1G', 'HLA-DRB1', 'CD74', 'LAPTM5', 'SRGN', 'GLUL', 'CTSS']\n",
      "B cell_lung cancer: ['CD79A', 'CD74', 'HLA-DRA', 'CXCR4', 'TTC19', 'MS4A1', 'CD37', 'HLA-DRB1']\n",
      "malignant cell_lung cancer: ['KRT19', 'KRT8', 'KRT18', 'PERP', 'SPINT2', 'EPCAM', 'MGST1', 'TXN', 'KRT7', 'TACSTD2']\n",
      "epithelial cell_lung cancer: ['SFTPB', 'SFTPA2', 'NAPSA', 'SFTPA1', 'KRT19', 'SFTA2', 'HOPX', 'SCGB3A2']\n",
      "T cell_lung cancer: ['CXCR4', 'CD3D', 'TRBC2', 'CD2', 'SRGN', 'RGS1', 'IL32', 'TRBC1', 'PTPRC', 'CD7']\n",
      "mast cell_lung cancer: ['CPA3', 'TPSAB1', 'TPSB2', 'MS4A2', 'KIT', 'SLC18A2', 'GATA2', 'HPGDS', 'VWA5A', 'SRGN']\n",
      "fibroblast_lung cancer: ['DCN', 'COL1A2', 'SPARC', 'COL6A2', 'BGN', 'COL3A1', 'SERPING1', 'C1R', 'LUM', 'C1S']\n",
      "plasmacytoid dendritic cell_lung cancer: ['GZMB', 'GPR183', 'JCHAIN', 'PLAC8', 'IRF7', 'IRF4', 'CD74', 'RGS2', 'TCF4', 'PLD4']\n",
      "neutrophil_lung cancer: ['G0S2', 'CXCL8', 'NAMPT', 'ALOX5AP', 'SRGN', 'BCL2A1', 'PLAUR', 'LITAF']\n",
      "epithelial cell_breast cancer: ['MUCL1', 'SCGB2A2', 'IGKV3-20', 'IGHV3-23', 'IGKV4-1', 'IGLV1-51', 'KRT19', 'MGP', 'AZGP1', 'IGHG1']\n",
      "endothelial cell_breast cancer: ['RAMP2', 'GNG11', 'EGFL7', 'PLVAP', 'SPARCL1', 'PECAM1', 'CRIP2', 'CD34', 'ADGRL4', 'IGFBP7']\n",
      "T cell_breast cancer: ['CD3E', 'IFITM1', 'RAC2', 'CORO1A', 'PTPRC', 'CD52', 'IL32']\n",
      "fibroblast_breast cancer: ['COL1A2', 'COL1A1', 'COL3A1', 'SPARC', 'COL6A2', 'COL6A1', 'CALD1', 'COL6A3', 'BGN', 'AEBP1']\n",
      "B cell_breast cancer: ['CD79A', 'CD74', 'MS4A1', 'NCF1', 'CD37', 'HLA-DRB5', 'HLA-DQA1']\n",
      "malignant cell_breast cancer: ['MIF', 'KRT19', 'PPDPF', 'COX6C', 'GABARAP', 'H3-3A', 'RPL30', 'TCEA1', 'RPL8']\n",
      "mast cell_breast cancer: ['TPSAB1', 'CPA3', 'TPSB2', 'MS4A2', 'GATA2', 'HPGDS', 'VWA5A', 'SLC18A2', 'LTC4S', 'HDC']\n",
      "mononuclear phagocyte_breast cancer: ['TYROBP', 'FCER1G', 'GPX1', 'AIF1', 'SPI1', 'CD74', 'GABARAP', 'CD68', 'CST3', 'HLA-DRB1']\n",
      "plasmacytoid dendritic cell_breast cancer: ['JCHAIN', 'IRF8', 'IRF7', 'LILRA4', 'PLAC8', 'GZMB', 'ITM2C', 'PLD4', 'CD74', 'CCDC50']\n",
      "T cell_melanoma: ['ASTN2', 'UGDH-AS1', 'SHISA9', 'KCNQ1OT1', 'IKZF3', 'ORC4', 'CFLAR', 'TMEM212', 'ANKRD20A9P', 'B2M']\n",
      "B cell_melanoma: ['CD74', 'ORAI2', 'UGDH-AS1', 'KCNQ1OT1', 'CFLAR', 'SHISA9', 'ORC4', 'TMEM212', 'ASTN2', 'METTL21A']\n",
      "plasmacytoid dendritic cell_melanoma: ['IRF7', 'MPEG1', 'PLD4', 'PLEK', 'IRF8', 'JCHAIN', 'PLAC8', 'GZMB']\n",
      "mononuclear phagocyte_melanoma: ['HLA-B', 'SAT1', 'HLA-C', 'SRGN', 'B2M', 'HLA-DRA', 'TAPBP', 'LYZ', 'METTL21A', 'ACTB']\n",
      "malignant cell_melanoma: ['GAPDH', 'HSP90AB1', 'KCNQ1OT1', 'RPL17', 'RPS17', 'ACTG1', 'EEF1G', 'ALDOA', 'EEF1A1', 'HSP90AA1']\n",
      "epithelial cell_melanoma: ['KCNQ1OT1', 'ANKRD20A9P', 'TOR1AIP2', 'MBOAT1', 'CFLAR', 'UGDH-AS1', 'CHP1', 'GAPDH', 'ORAI2', 'GATD1']\n",
      "fibroblast_melanoma: ['EEF1G', 'KCNQ1OT1', 'UGDH-AS1', 'VIM', 'CHP1', 'RPL13AP5', 'ACTB', 'EEF1A1', 'RPS17', 'ACTG1']\n",
      "endothelial cell_melanoma: ['SHISA9', 'KCNQ1OT1', 'ANKRD20A9P', 'ORC4', 'EEF1G', 'VIM', 'CFLAR', 'TNFAIP8L1', 'HLA-B', 'RBMS2']\n",
      "mast cell_melanoma: ['TPSD1', 'CPA3', 'TPSAB1', 'TPSB2', 'MS4A2', 'HPGDS', 'CTSG', 'SLC18A2', 'VWA5A']\n",
      "T cell_liver cancer: ['RPS29', 'LTB', 'RPS27', 'RPLP2', 'CD3D', 'RPS10', 'CD52', 'RPL21', 'HBB', 'RPL39']\n",
      "endothelial cell_liver cancer: ['RAMP2', 'EGFL7', 'HSPG2', 'GNG11', 'PLVAP', 'ENG', 'PECAM1', 'TCF4', 'FKBP1A', 'CRIP2']\n",
      "fibroblast_liver cancer: ['IGFBP7', 'CALD1', 'BGN', 'MYL9', 'RGS5', 'PPP1R14A', 'MGP', 'TAGLN', 'PLAC9', 'SPARC']\n",
      "epithelial cell_liver cancer: ['ALB', 'ATP1B1', 'SPP1', 'VTN', 'FXYD2', 'TM4SF4', 'PDZK1IP1', 'ANXA4', 'CLDN10', 'SERPINA1']\n",
      "malignant cell_liver cancer: ['KRT18', 'UQCRQ', 'SEC61G', 'KRT8', 'SERPINA1', 'ATP5MF', 'TM4SF4', 'ELOB', 'COX6A1']\n",
      "B cell_liver cancer: ['CD79A', 'RPS29', 'RPS10', 'IGKC', 'CD74', 'RPLP2', 'RPS27', 'MS4A1', 'HBB', 'RPL27A']\n",
      "plasmacytoid dendritic cell_liver cancer: ['JCHAIN', 'IRF7', 'GZMB', 'PLAC8', 'GPR183', 'PPP1R14B', 'IRF8', 'RGS2', 'ITM2C', 'SEC61B']\n",
      "mononuclear phagocyte_liver cancer: ['AIF1', 'HLA-DRA', 'TYROBP', 'LST1', 'HLA-DPB1', 'HLA-DRB1', 'CST3', 'CD74', 'HLA-DPA1', 'LYZ']\n",
      "mast cell_liver cancer: ['TPSAB1', 'TPSB2', 'HPGDS', 'IL1RL1', 'CPA3', 'LINC00623', 'REST', 'ALOX5AP', 'SRGN', 'ID2']\n",
      "neutrophil_uveal melanoma: ['FCGR3B', 'CSF3R', 'S100A8', 'MNDA', 'S100A9', 'NAMPT', 'RGS2', 'FPR1', 'RIPOR2', 'MXD1']\n",
      "T cell_uveal melanoma: ['CST7', 'CD3E', 'CREM', 'CD3D', 'PTPRC', 'SRGN', 'IL32', 'TNFAIP3', 'NKG7']\n",
      "endothelial cell_uveal melanoma: ['GNG11', 'RAMP2', 'CAVIN2', 'IFITM3', 'PECAM1', 'TM4SF1', 'RAMP3', 'CRIP2', 'ESAM', 'ENG']\n",
      "B cell_uveal melanoma: ['IGHG1', 'CD79A', 'JCHAIN', 'HERPUD1', 'MZB1', 'SSR4']\n",
      "mononuclear phagocyte_uveal melanoma: ['TYROBP', 'CST3', 'CD74', 'AIF1', 'FCER1G', 'FTL', 'HLA-DRA', 'HLA-DPA1', 'CD68', 'MS4A7']\n",
      "plasmacytoid dendritic cell_uveal melanoma: ['GZMB', 'LILRA4', 'JCHAIN', 'PLD4', 'IRF8', 'IRF7', 'SEC61B', 'CD74', 'PLAC8', 'GPR183']\n",
      "fibroblast_uveal melanoma: ['FRZB', 'CST3', 'TIMP3', 'IGFBP7', 'SELENOW', 'RBP1', 'SERPINF1', 'PTGDS']\n",
      "fibroblast_colorectal cancer: ['CALD1', 'LGALS1', 'TPM2', 'IGFBP7', 'SELENOM', 'COL3A1', 'SPARC', 'TIMP1', 'COL1A2', 'COL6A2']\n",
      "endothelial cell_colorectal cancer: ['RAMP2', 'GNG11', 'PECAM1', 'FKBP1A', 'IGFBP7', 'CRIP2', 'TM4SF1', 'HSPG2', 'VWF', 'SPARCL1']\n",
      "mononuclear phagocyte_colorectal cancer: ['TYROBP', 'FCER1G', 'AIF1', 'LYZ', 'CYBA', 'FTL', 'GPX1', 'HLA-DRA', 'SRGN', 'LST1']\n",
      "B cell_colorectal cancer: ['CD79A', 'MS4A1', 'IGKC', 'CD37', 'RPL21', 'CD52', 'RPS29', 'CD79B', 'HLA-DRA', 'RPLP2']\n",
      "epithelial cell_colorectal cancer: ['PHGR1', 'TFF3', 'LGALS4', 'KRT8', 'S100A6', 'KRT18', 'FXYD3', 'TSPAN8', 'S100P', 'AGR2']\n",
      "T cell_colorectal cancer: ['CD3D', 'CD2', 'IL32', 'CD52', 'CORO1A', 'TMSB4X', 'RPS29', 'BTG1', 'MALAT1', 'CD3E']\n",
      "mast cell_colorectal cancer: ['TPSAB1', 'CPA3', 'TPSB2', 'MS4A2', 'HPGDS', 'SAMSN1', 'LTC4S', 'HPGD', 'RGS1', 'ALOX5AP']\n",
      "plasmacytoid dendritic cell_colorectal cancer: ['JCHAIN', 'GZMB', 'IRF8', 'PLAC8', 'IRF7', 'SEC61B', 'UGCG', 'IRF4', 'CCDC50', 'CLN8']\n",
      "neutrophil_colorectal cancer: ['CXCL8', 'BCL2A1', 'S100A8', 'NAMPT', 'G0S2', 'S100A9', 'SAT1', 'FTH1', 'ALOX5AP', 'PLEK']\n",
      "malignant cell_colorectal cancer: ['PHGR1', 'KRT8', 'S100A6', 'TFF3', 'LGALS4', 'KRT18', 'TSPAN8', 'EPCAM', 'KRT19', 'C19orf33']\n",
      "epithelial cell_ovarian cancer: ['HBB', 'HBA2', 'HBA1', 'WFDC2', 'COL1A1', 'COL1A2', 'TAGLN', 'COL3A1', 'RBP1', 'FABP4']\n",
      "fibroblast_ovarian cancer: ['COL1A2', 'COL1A1', 'DCN', 'CALD1', 'COL3A1', 'LUM', 'SPARC', 'LGALS1', 'RARRES2', 'TPM2']\n",
      "mononuclear phagocyte_ovarian cancer: ['TYROBP', 'AIF1', 'HLA-DRA', 'FCER1G', 'HLA-DPB1', 'HLA-DPA1', 'HLA-DRB1', 'CD74', 'FTL', 'SAT1']\n",
      "malignant cell_ovarian cancer: ['KRT18', 'KRT8', 'CD24', 'WFDC2', 'RPL7', 'GSTP1', 'MDK', 'KRT19', 'YWHAE', 'RBP1']\n",
      "T cell_ovarian cancer: ['CXCR4', 'IL32', 'BTG1', 'CD2', 'CD3D', 'CD52', 'CORO1A', 'GZMA', 'IGKC', 'TRBC1']\n",
      "endothelial cell_ovarian cancer: ['SPARCL1', 'GNG11', 'RAMP2', 'VWF', 'IGFBP7', 'CLDN5', 'PECAM1', 'TM4SF1', 'HSPG2', 'CRIP2']\n",
      "B cell_ovarian cancer: ['IGKC', 'IGLC2', 'IGHG3', 'IGHG1', 'IGHG4', 'CD79A', 'IGLC3', 'IGHGP', 'HLA-DRA', 'MZB1']\n",
      "mast cell_ovarian cancer: ['TPSAB1', 'TPSB2', 'CPA3', 'HPGDS', 'MS4A2', 'RGS13', 'VWA5A', 'FCER1G', 'ALOX5AP', 'LTC4S']\n",
      "plasmacytoid dendritic cell_ovarian cancer: ['GZMB', 'JCHAIN', 'GPR183', 'IRF7', 'PLAC8', 'ALOX5AP', 'AREG', 'C12orf75', 'TSPAN13', 'CD74']\n",
      "neutrophil_ovarian cancer: ['CXCL8', 'BCL2A1', 'C15orf48', 'PELATON', 'S100A9', 'G0S2', 'NAMPT', 'FTH1', 'MXD1', 'S100A8']\n",
      "epithelial cell_uveal melanoma: ['MLANA', 'TYRP1', 'PMEL', 'S100A1', 'SNHG7', 'MITF', 'TEX41', 'SLCO4A1-AS1', 'RAB38']\n",
      "malignant cell_uveal melanoma: ['S100A1', 'CITED1', 'SDCBP', 'MLANA', 'ZNF106', 'RAB38', 'MIF', 'EFHD1', 'ANXA5', 'ST3GAL4']\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------\n",
    "# Final: Print marker genes per cell type_disease group\n",
    "# -----------------------------\n",
    "filtered_marker_dict = {}\n",
    "\n",
    "for group, genes in converted_marker_genes_by_group.items():\n",
    "    valid_genes = [g for g in genes if g in adata.var[\"feature_name\"].values]\n",
    "    if valid_genes:\n",
    "        filtered_marker_dict[group] = valid_genes\n",
    "\n",
    "# Print in requested format\n",
    "for group, genes in filtered_marker_dict.items():\n",
    "    print(f\"{group}: {genes}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2b852b89-0b14-4b69-af07-7410338b19e2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "endothelial cell_lung cancer: ['VWF', 'EPAS1', 'PECAM1', 'CLDN5', 'CLEC14A', 'SPARCL1', 'RAMP2', 'HYAL2', 'AQP1', 'TM4SF1']\n",
      "endothelial cell_breast cancer: ['RAMP2', 'GNG11', 'EGFL7', 'PLVAP', 'SPARCL1', 'PECAM1', 'CRIP2', 'CD34', 'ADGRL4', 'IGFBP7']\n",
      "endothelial cell_melanoma: ['SHISA9', 'KCNQ1OT1', 'ANKRD20A9P', 'ORC4', 'EEF1G', 'VIM', 'CFLAR', 'TNFAIP8L1', 'HLA-B', 'RBMS2']\n",
      "endothelial cell_liver cancer: ['RAMP2', 'EGFL7', 'HSPG2', 'GNG11', 'PLVAP', 'ENG', 'PECAM1', 'TCF4', 'FKBP1A', 'CRIP2']\n",
      "endothelial cell_uveal melanoma: ['GNG11', 'RAMP2', 'CAVIN2', 'IFITM3', 'PECAM1', 'TM4SF1', 'RAMP3', 'CRIP2', 'ESAM', 'ENG']\n",
      "endothelial cell_colorectal cancer: ['RAMP2', 'GNG11', 'PECAM1', 'FKBP1A', 'IGFBP7', 'CRIP2', 'TM4SF1', 'HSPG2', 'VWF', 'SPARCL1']\n",
      "endothelial cell_ovarian cancer: ['SPARCL1', 'GNG11', 'RAMP2', 'VWF', 'IGFBP7', 'CLDN5', 'PECAM1', 'TM4SF1', 'HSPG2', 'CRIP2']\n",
      "mononuclear phagocyte_lung cancer: ['TYROBP', 'LYZ', 'HLA-DRA', 'FCER1G', 'HLA-DRB1', 'CD74', 'LAPTM5', 'SRGN', 'GLUL', 'CTSS']\n",
      "mononuclear phagocyte_breast cancer: ['TYROBP', 'FCER1G', 'GPX1', 'AIF1', 'SPI1', 'CD74', 'GABARAP', 'CD68', 'CST3', 'HLA-DRB1']\n",
      "mononuclear phagocyte_melanoma: ['HLA-B', 'SAT1', 'HLA-C', 'SRGN', 'B2M', 'HLA-DRA', 'TAPBP', 'LYZ', 'METTL21A', 'ACTB']\n",
      "mononuclear phagocyte_liver cancer: ['AIF1', 'HLA-DRA', 'TYROBP', 'LST1', 'HLA-DPB1', 'HLA-DRB1', 'CST3', 'CD74', 'HLA-DPA1', 'LYZ']\n",
      "mononuclear phagocyte_uveal melanoma: ['TYROBP', 'CST3', 'CD74', 'AIF1', 'FCER1G', 'FTL', 'HLA-DRA', 'HLA-DPA1', 'CD68', 'MS4A7']\n",
      "mononuclear phagocyte_colorectal cancer: ['TYROBP', 'FCER1G', 'AIF1', 'LYZ', 'CYBA', 'FTL', 'GPX1', 'HLA-DRA', 'SRGN', 'LST1']\n",
      "mononuclear phagocyte_ovarian cancer: ['TYROBP', 'AIF1', 'HLA-DRA', 'FCER1G', 'HLA-DPB1', 'HLA-DPA1', 'HLA-DRB1', 'CD74', 'FTL', 'SAT1']\n",
      "B cell_lung cancer: ['CD79A', 'CD74', 'HLA-DRA', 'CXCR4', 'TTC19', 'MS4A1', 'CD37', 'HLA-DRB1']\n",
      "B cell_breast cancer: ['CD79A', 'CD74', 'MS4A1', 'NCF1', 'CD37', 'HLA-DRB5', 'HLA-DQA1']\n",
      "B cell_melanoma: ['CD74', 'ORAI2', 'UGDH-AS1', 'KCNQ1OT1', 'CFLAR', 'SHISA9', 'ORC4', 'TMEM212', 'ASTN2', 'METTL21A']\n",
      "B cell_liver cancer: ['CD79A', 'RPS29', 'RPS10', 'IGKC', 'CD74', 'RPLP2', 'RPS27', 'MS4A1', 'HBB', 'RPL27A']\n",
      "B cell_uveal melanoma: ['IGHG1', 'CD79A', 'JCHAIN', 'HERPUD1', 'MZB1', 'SSR4']\n",
      "B cell_colorectal cancer: ['CD79A', 'MS4A1', 'IGKC', 'CD37', 'RPL21', 'CD52', 'RPS29', 'CD79B', 'HLA-DRA', 'RPLP2']\n",
      "B cell_ovarian cancer: ['IGKC', 'IGLC2', 'IGHG3', 'IGHG1', 'IGHG4', 'CD79A', 'IGLC3', 'IGHGP', 'HLA-DRA', 'MZB1']\n",
      "malignant cell_lung cancer: ['KRT19', 'KRT8', 'KRT18', 'PERP', 'SPINT2', 'EPCAM', 'MGST1', 'TXN', 'KRT7', 'TACSTD2']\n",
      "malignant cell_breast cancer: ['MIF', 'KRT19', 'PPDPF', 'COX6C', 'GABARAP', 'H3-3A', 'RPL30', 'TCEA1', 'RPL8']\n",
      "malignant cell_melanoma: ['GAPDH', 'HSP90AB1', 'KCNQ1OT1', 'RPL17', 'RPS17', 'ACTG1', 'EEF1G', 'ALDOA', 'EEF1A1', 'HSP90AA1']\n",
      "malignant cell_liver cancer: ['KRT18', 'UQCRQ', 'SEC61G', 'KRT8', 'SERPINA1', 'ATP5MF', 'TM4SF4', 'ELOB', 'COX6A1']\n",
      "malignant cell_colorectal cancer: ['PHGR1', 'KRT8', 'S100A6', 'TFF3', 'LGALS4', 'KRT18', 'TSPAN8', 'EPCAM', 'KRT19', 'C19orf33']\n",
      "malignant cell_ovarian cancer: ['KRT18', 'KRT8', 'CD24', 'WFDC2', 'RPL7', 'GSTP1', 'MDK', 'KRT19', 'YWHAE', 'RBP1']\n",
      "malignant cell_uveal melanoma: ['S100A1', 'CITED1', 'SDCBP', 'MLANA', 'ZNF106', 'RAB38', 'MIF', 'EFHD1', 'ANXA5', 'ST3GAL4']\n",
      "epithelial cell_lung cancer: ['SFTPB', 'SFTPA2', 'NAPSA', 'SFTPA1', 'KRT19', 'SFTA2', 'HOPX', 'SCGB3A2']\n",
      "epithelial cell_breast cancer: ['MUCL1', 'SCGB2A2', 'IGKV3-20', 'IGHV3-23', 'IGKV4-1', 'IGLV1-51', 'KRT19', 'MGP', 'AZGP1', 'IGHG1']\n",
      "epithelial cell_melanoma: ['KCNQ1OT1', 'ANKRD20A9P', 'TOR1AIP2', 'MBOAT1', 'CFLAR', 'UGDH-AS1', 'CHP1', 'GAPDH', 'ORAI2', 'GATD1']\n",
      "epithelial cell_liver cancer: ['ALB', 'ATP1B1', 'SPP1', 'VTN', 'FXYD2', 'TM4SF4', 'PDZK1IP1', 'ANXA4', 'CLDN10', 'SERPINA1']\n",
      "epithelial cell_colorectal cancer: ['PHGR1', 'TFF3', 'LGALS4', 'KRT8', 'S100A6', 'KRT18', 'FXYD3', 'TSPAN8', 'S100P', 'AGR2']\n",
      "epithelial cell_ovarian cancer: ['HBB', 'HBA2', 'HBA1', 'WFDC2', 'COL1A1', 'COL1A2', 'TAGLN', 'COL3A1', 'RBP1', 'FABP4']\n",
      "epithelial cell_uveal melanoma: ['MLANA', 'TYRP1', 'PMEL', 'S100A1', 'SNHG7', 'MITF', 'TEX41', 'SLCO4A1-AS1', 'RAB38']\n",
      "T cell_lung cancer: ['CXCR4', 'CD3D', 'TRBC2', 'CD2', 'SRGN', 'RGS1', 'IL32', 'TRBC1', 'PTPRC', 'CD7']\n",
      "T cell_breast cancer: ['CD3E', 'IFITM1', 'RAC2', 'CORO1A', 'PTPRC', 'CD52', 'IL32']\n",
      "T cell_melanoma: ['ASTN2', 'UGDH-AS1', 'SHISA9', 'KCNQ1OT1', 'IKZF3', 'ORC4', 'CFLAR', 'TMEM212', 'ANKRD20A9P', 'B2M']\n",
      "T cell_liver cancer: ['RPS29', 'LTB', 'RPS27', 'RPLP2', 'CD3D', 'RPS10', 'CD52', 'RPL21', 'HBB', 'RPL39']\n",
      "T cell_uveal melanoma: ['CST7', 'CD3E', 'CREM', 'CD3D', 'PTPRC', 'SRGN', 'IL32', 'TNFAIP3', 'NKG7']\n",
      "T cell_colorectal cancer: ['CD3D', 'CD2', 'IL32', 'CD52', 'CORO1A', 'TMSB4X', 'RPS29', 'BTG1', 'MALAT1', 'CD3E']\n",
      "T cell_ovarian cancer: ['CXCR4', 'IL32', 'BTG1', 'CD2', 'CD3D', 'CD52', 'CORO1A', 'GZMA', 'IGKC', 'TRBC1']\n",
      "mast cell_lung cancer: ['CPA3', 'TPSAB1', 'TPSB2', 'MS4A2', 'KIT', 'SLC18A2', 'GATA2', 'HPGDS', 'VWA5A', 'SRGN']\n",
      "mast cell_breast cancer: ['TPSAB1', 'CPA3', 'TPSB2', 'MS4A2', 'GATA2', 'HPGDS', 'VWA5A', 'SLC18A2', 'LTC4S', 'HDC']\n",
      "mast cell_melanoma: ['TPSD1', 'CPA3', 'TPSAB1', 'TPSB2', 'MS4A2', 'HPGDS', 'CTSG', 'SLC18A2', 'VWA5A']\n",
      "mast cell_liver cancer: ['TPSAB1', 'TPSB2', 'HPGDS', 'IL1RL1', 'CPA3', 'LINC00623', 'REST', 'ALOX5AP', 'SRGN', 'ID2']\n",
      "mast cell_colorectal cancer: ['TPSAB1', 'CPA3', 'TPSB2', 'MS4A2', 'HPGDS', 'SAMSN1', 'LTC4S', 'HPGD', 'RGS1', 'ALOX5AP']\n",
      "mast cell_ovarian cancer: ['TPSAB1', 'TPSB2', 'CPA3', 'HPGDS', 'MS4A2', 'RGS13', 'VWA5A', 'FCER1G', 'ALOX5AP', 'LTC4S']\n",
      "fibroblast_lung cancer: ['DCN', 'COL1A2', 'SPARC', 'COL6A2', 'BGN', 'COL3A1', 'SERPING1', 'C1R', 'LUM', 'C1S']\n",
      "fibroblast_breast cancer: ['COL1A2', 'COL1A1', 'COL3A1', 'SPARC', 'COL6A2', 'COL6A1', 'CALD1', 'COL6A3', 'BGN', 'AEBP1']\n",
      "fibroblast_melanoma: ['EEF1G', 'KCNQ1OT1', 'UGDH-AS1', 'VIM', 'CHP1', 'RPL13AP5', 'ACTB', 'EEF1A1', 'RPS17', 'ACTG1']\n",
      "fibroblast_liver cancer: ['IGFBP7', 'CALD1', 'BGN', 'MYL9', 'RGS5', 'PPP1R14A', 'MGP', 'TAGLN', 'PLAC9', 'SPARC']\n",
      "fibroblast_uveal melanoma: ['FRZB', 'CST3', 'TIMP3', 'IGFBP7', 'SELENOW', 'RBP1', 'SERPINF1', 'PTGDS']\n",
      "fibroblast_colorectal cancer: ['CALD1', 'LGALS1', 'TPM2', 'IGFBP7', 'SELENOM', 'COL3A1', 'SPARC', 'TIMP1', 'COL1A2', 'COL6A2']\n",
      "fibroblast_ovarian cancer: ['COL1A2', 'COL1A1', 'DCN', 'CALD1', 'COL3A1', 'LUM', 'SPARC', 'LGALS1', 'RARRES2', 'TPM2']\n",
      "plasmacytoid dendritic cell_lung cancer: ['GZMB', 'GPR183', 'JCHAIN', 'PLAC8', 'IRF7', 'IRF4', 'CD74', 'RGS2', 'TCF4', 'PLD4']\n",
      "plasmacytoid dendritic cell_breast cancer: ['JCHAIN', 'IRF8', 'IRF7', 'LILRA4', 'PLAC8', 'GZMB', 'ITM2C', 'PLD4', 'CD74', 'CCDC50']\n",
      "plasmacytoid dendritic cell_melanoma: ['IRF7', 'MPEG1', 'PLD4', 'PLEK', 'IRF8', 'JCHAIN', 'PLAC8', 'GZMB']\n",
      "plasmacytoid dendritic cell_liver cancer: ['JCHAIN', 'IRF7', 'GZMB', 'PLAC8', 'GPR183', 'PPP1R14B', 'IRF8', 'RGS2', 'ITM2C', 'SEC61B']\n",
      "plasmacytoid dendritic cell_uveal melanoma: ['GZMB', 'LILRA4', 'JCHAIN', 'PLD4', 'IRF8', 'IRF7', 'SEC61B', 'CD74', 'PLAC8', 'GPR183']\n",
      "plasmacytoid dendritic cell_colorectal cancer: ['JCHAIN', 'GZMB', 'IRF8', 'PLAC8', 'IRF7', 'SEC61B', 'UGCG', 'IRF4', 'CCDC50', 'CLN8']\n",
      "plasmacytoid dendritic cell_ovarian cancer: ['GZMB', 'JCHAIN', 'GPR183', 'IRF7', 'PLAC8', 'ALOX5AP', 'AREG', 'C12orf75', 'TSPAN13', 'CD74']\n",
      "neutrophil_lung cancer: ['G0S2', 'CXCL8', 'NAMPT', 'ALOX5AP', 'SRGN', 'BCL2A1', 'PLAUR', 'LITAF']\n",
      "neutrophil_uveal melanoma: ['FCGR3B', 'CSF3R', 'S100A8', 'MNDA', 'S100A9', 'NAMPT', 'RGS2', 'FPR1', 'RIPOR2', 'MXD1']\n",
      "neutrophil_colorectal cancer: ['CXCL8', 'BCL2A1', 'S100A8', 'NAMPT', 'G0S2', 'S100A9', 'SAT1', 'FTH1', 'ALOX5AP', 'PLEK']\n",
      "neutrophil_ovarian cancer: ['CXCL8', 'BCL2A1', 'C15orf48', 'PELATON', 'S100A9', 'G0S2', 'NAMPT', 'FTH1', 'MXD1', 'S100A8']\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------\n",
    "# Final: Print marker genes grouped by cell type and cancer\n",
    "# -----------------------------\n",
    "from collections import defaultdict\n",
    "\n",
    "grouped_by_celltype = defaultdict(dict)\n",
    "\n",
    "for group, genes in converted_marker_genes_by_group.items():\n",
    "    if \"_\" in group:\n",
    "        cell_type, disease = group.split(\"_\", 1)\n",
    "        grouped_by_celltype[cell_type][disease] = [g for g in genes if g in adata.var[\"feature_name\"].values]\n",
    "\n",
    "# Print in requested format\n",
    "for cell_type, disease_dict in grouped_by_celltype.items():\n",
    "    for disease, genes in disease_dict.items():\n",
    "        print(f\"{cell_type}_{disease}: {genes}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eb92c9f0-9f57-42a5-aab8-d607ed1f6dcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes in adata: 2188\n",
      "Included marker genes: 300\n",
      "HVGs in adata: 2000\n"
     ]
    }
   ],
   "source": [
    "# Total genes in adata\n",
    "print(f\"Genes in adata: {adata.shape[1]}\")\n",
    "\n",
    "# Marker genes in adata\n",
    "print(f\"Included marker genes: {len(marker_genes_in_data)}\")\n",
    "\n",
    "# HVG genes in adata\n",
    "print(f\"HVGs in adata: {adata.var['highly_variable'].sum()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8bfb0b27-b538-418d-8614-050cc1c7165a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Included marker genes in adata: 300 / 311\n",
      "Some examples: ['GLUL', 'MIF', 'ENG', 'ESAM', 'TPSAB1', 'CLN8', 'ORAI2', 'CXCR4', 'CD52', 'PPP1R14A']\n"
     ]
    }
   ],
   "source": [
    "# This was your confirmed list of marker genes\n",
    "included_marker_genes = [g for g in all_marker_genes if g in adata.var[\"feature_name\"].values]\n",
    "\n",
    "print(f\"Included marker genes in adata: {len(included_marker_genes)} / {len(all_marker_genes)}\")\n",
    "print(\"Some examples:\", included_marker_genes[:10])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a4253dbf-a2ce-4935-97ca-50b53eacbc6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes used for scGPT training: 2188\n"
     ]
    }
   ],
   "source": [
    "print(\"Genes used for scGPT training:\", len(adata.var[\"feature_name\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a39b0778-3921-4af6-a5fb-20b3f3ef4e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"adata_hvg_marker.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0245bddb-eee7-4181-b3ca-f24ed28f30bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 234610 × 2188\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'harm_study', 'harm_healthy.tissue', 'harm_tumor.site', 'harm_sample.type', 'harm_condition', 'harm_tumor.type', 'harm_cd45pos', 'harm_healthy.pat', 'percent.mt', 'ratio_nCount_nFeature', 'batch', 'X_scvi_batch', 'X_scvi_labels', 'X_scvi_local_l_mean', 'X_scvi_local_l_var', 'leiden_0.2', 'leiden_0.4', 'leiden_0.6', 'leiden_0.8', 'leiden_1', 'leiden_1.2', 'leiden_1.4', 'author_first_cell_type', 'author_cell_type', 'cnv_score', 'organism_ontology_term_id', 'donor_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'suspension_type', 'assay_ontology_term_id', 'cell_id', 'dup', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'group'\n",
      "    var: 'features', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
      "    uns: 'citation', 'hvg', 'log1p', 'schema_reference', 'schema_version', 'title'\n",
      "    obsm: 'X_umap'\n"
     ]
    }
   ],
   "source": [
    "import anndata\n",
    "\n",
    "# Load the .h5ad file\n",
    "adata = anndata.read_h5ad(\"adata_hvg_marker.h5ad\")\n",
    "\n",
    "# Check basic details about the dataset\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "304f65c6-3c9b-4b82-b605-f0c1b5fa4cee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['endothelial cell_lung cancer', 'mononuclear phagocyte_lung cancer', 'B cell_lung cancer', 'malignant cell_lung cancer', 'epithelial cell_lung cancer', ..., 'mast cell_ovarian cancer', 'plasmacytoid dendritic cell_ovarian cancer', 'neutrophil_ovarian cancer', 'epithelial cell_uveal melanoma', 'malignant cell_uveal melanoma']\n",
      "Length: 66\n",
      "Categories (66, object): ['B cell_breast cancer', 'B cell_colorectal cancer', 'B cell_liver cancer', 'B cell_lung cancer', ..., 'plasmacytoid dendritic cell_lung cancer', 'plasmacytoid dendritic cell_melanoma', 'plasmacytoid dendritic cell_ovarian cancer', 'plasmacytoid dendritic cell_uveal melanoma']\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs[\"group\"].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c94bb600-2093-4c56-9fe3-66f89bf46e30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                          cell_type      disease  \\\n",
      "BT1299_GTACTCCTCGGAAACG-1-17       endothelial cell  lung cancer   \n",
      "BT1299_GTACTTTAGCAGGCTA-1-17  mononuclear phagocyte  lung cancer   \n",
      "BT1299_GTACTTTAGCATGGCA-1-17       endothelial cell  lung cancer   \n",
      "BT1299_GTACTTTAGGATGTAT-1-17                 B cell  lung cancer   \n",
      "BT1299_GTACTTTAGGCTCTTA-1-17         malignant cell  lung cancer   \n",
      "\n",
      "                                                          group  \n",
      "BT1299_GTACTCCTCGGAAACG-1-17       endothelial cell_lung cancer  \n",
      "BT1299_GTACTTTAGCAGGCTA-1-17  mononuclear phagocyte_lung cancer  \n",
      "BT1299_GTACTTTAGCATGGCA-1-17       endothelial cell_lung cancer  \n",
      "BT1299_GTACTTTAGGATGTAT-1-17                 B cell_lung cancer  \n",
      "BT1299_GTACTTTAGGCTCTTA-1-17         malignant cell_lung cancer  \n"
     ]
    }
   ],
   "source": [
    "print(adata.obs[[\"cell_type\", \"disease\", \"group\"]].head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9b398738-aa07-42c0-a25c-96f212575be6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "group\n",
      "T cell_lung cancer                               30482\n",
      "epithelial cell_uveal melanoma                   23628\n",
      "malignant cell_uveal melanoma                    18137\n",
      "fibroblast_ovarian cancer                        15163\n",
      "mononuclear phagocyte_lung cancer                13364\n",
      "                                                 ...  \n",
      "plasmacytoid dendritic cell_colorectal cancer       24\n",
      "neutrophil_ovarian cancer                           11\n",
      "neutrophil_uveal melanoma                           11\n",
      "mast cell_liver cancer                               4\n",
      "mast cell_melanoma                                   3\n",
      "Name: count, Length: 66, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs[\"group\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9e49cfae-50b2-405c-909e-5d5c3468252a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "print(\"group\" in adata.obs.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "42289402-6c1a-4e18-8341-26a486ea7ea7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['orig.ident', 'nCount_RNA', 'nFeature_RNA', 'harm_study', 'harm_healthy.tissue', 'harm_tumor.site', 'harm_sample.type', 'harm_condition', 'harm_tumor.type', 'harm_cd45pos', 'harm_healthy.pat', 'percent.mt', 'ratio_nCount_nFeature', 'batch', 'X_scvi_batch', 'X_scvi_labels', 'X_scvi_local_l_mean', 'X_scvi_local_l_var', 'leiden_0.2', 'leiden_0.4', 'leiden_0.6', 'leiden_0.8', 'leiden_1', 'leiden_1.2', 'leiden_1.4', 'author_first_cell_type', 'author_cell_type', 'cnv_score', 'organism_ontology_term_id', 'donor_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'suspension_type', 'assay_ontology_term_id', 'cell_id', 'dup', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'group']\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "23304192-d22c-4228-b01c-710d64e046e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cell_type\n",
      "T cell                         65153\n",
      "malignant cell                 47442\n",
      "epithelial cell                38933\n",
      "mononuclear phagocyte          27218\n",
      "fibroblast                     26306\n",
      "B cell                         16527\n",
      "endothelial cell                7255\n",
      "neutrophil                      2824\n",
      "mast cell                       1989\n",
      "plasmacytoid dendritic cell      963\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs[\"cell_type\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6bad4f1e-5c43-402c-9894-5980026ce2d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train dataset: 187688 cells\n",
      "Test dataset: 46922 cells\n",
      "Loaded train dataset: (187688, 2188)\n",
      "Loaded test dataset: (46922, 2188)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import anndata as ad\n",
    "\n",
    "# Split with stratification\n",
    "train_indices, test_indices = train_test_split(\n",
    "    adata.obs_names,\n",
    "    test_size=0.2,\n",
    "    stratify=adata.obs[\"group\"],\n",
    "    random_state=42,\n",
    ")\n",
    "\n",
    "# Subset\n",
    "train_dataset = adata[train_indices].copy()\n",
    "test_dataset = adata[test_indices].copy()\n",
    "\n",
    "# Save\n",
    "train_dataset.write_h5ad(\"train_datamarker.h5ad\")\n",
    "test_dataset.write_h5ad(\"test_datamarker.h5ad\")\n",
    "\n",
    "print(f\"Train dataset: {train_dataset.shape[0]} cells\")\n",
    "print(f\"Test dataset: {test_dataset.shape[0]} cells\")\n",
    "\n",
    "# Reload if needed\n",
    "train_dataset = ad.read_h5ad(\"train_datamarker.h5ad\")\n",
    "test_dataset = ad.read_h5ad(\"test_datamarker.h5ad\")\n",
    "\n",
    "print(f\"Loaded train dataset: {train_dataset.shape}\")\n",
    "print(f\"Loaded test dataset: {test_dataset.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6ffa4678-4ee2-4e60-a9c9-f0229c5f26d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Helical and scGPT imported successfully!\n"
     ]
    }
   ],
   "source": [
    "import helical\n",
    "from helical.models.scgpt.model import scGPT, scGPTConfig\n",
    "\n",
    "\n",
    "print(\"Helical and scGPT imported successfully!\")\n",
    "\n",
    "from helical.models.scgpt import scGPTConfig, scGPTFineTuningModel  # ✅ Correct Import\n",
    "import torch\n",
    "import numpy as np\n",
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay, classification_report\n",
    "import matplotlib.pyplot as plt\n",
    "import logging, warnings\n",
    "import umap\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import anndata as ad "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "659e3515-8b65-47ee-9414-f8ace7a349bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Train dataset: 187688 cells\n",
      " Test dataset: 46922 cells\n",
      "Loaded train dataset: (187688, 2188)\n",
      "Loaded test dataset: (46922, 2188)\n",
      "Converted gene expression matrix to integer UMI counts.\n",
      "Found 66 unique group labels.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helical.models.scgpt.model:Model finished initializing.\n",
      "INFO:helical.models.scgpt.model:'scGPT' model is in 'eval' mode, on device 'cuda' with embedding mode 'cls'.\n",
      "INFO:helical.models.scgpt.model:Processing data for scGPT.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scGPT fine-tuning model initialized successfully.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/pandas/core/frame.py:6479: ImplicitModificationWarning: Trying to modify index of attribute `.var` of view, initializing view as actual.\n",
      "  new_obj.index = new_index\n",
      "\n",
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "\n",
      "INFO:helical.models.scgpt.model:Successfully processed the data for scGPT.\n",
      "INFO:helical.models.scgpt.model:Processing data for scGPT.\n",
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/pandas/core/frame.py:6479: ImplicitModificationWarning: Trying to modify index of attribute `.var` of view, initializing view as actual.\n",
      "  new_obj.index = new_index\n",
      "\n",
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "\n",
      "INFO:helical.models.scgpt.model:Successfully processed the data for scGPT.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed training dataset size: 187688\n",
      "Processed validation dataset size: 46922\n",
      "Encoded 66 unique groups.\n",
      "Log-scaled normalized class weights: tensor([0.8413, 0.9679, 1.1711, 0.5796, 0.9881, 0.8200, 1.0100, 0.5529, 0.7757,\n",
      "        0.8683, 0.3827, 1.0418, 0.5403, 0.6291, 0.9064, 0.9676, 0.9639, 0.8910,\n",
      "        1.3871, 0.8459, 1.2289, 0.7797, 0.7779, 0.9876, 0.7773, 1.2474, 0.6698,\n",
      "        0.4232, 0.6869, 0.8010, 0.9758, 0.8247, 1.3410, 0.4958, 1.1622, 0.7311,\n",
      "        1.0548, 0.9333, 0.5727, 0.8640, 0.5311, 0.4662, 1.2095, 1.2139, 1.9542,\n",
      "        0.9117, 2.0260, 1.3448, 0.8766, 0.9064, 1.0993, 0.5168, 1.5097, 0.5905,\n",
      "        0.8820, 1.4396, 0.7900, 1.7598, 1.7598, 1.2987, 1.6276, 1.4428, 1.1277,\n",
      "        1.5938, 1.2431, 1.4128], device='cuda:0')\n"
     ]
    }
   ],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "from torch.nn import CrossEntropyLoss\n",
    "from sklearn.model_selection import train_test_split\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "# **Set device (Use GPU if available)**\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "# **Load preprocessed train and test datasets**\n",
    "adata = ad.read_h5ad(\"adata_hvg_marker.h5ad\")\n",
    "\n",
    "# **Ensure 'group' column exists**\n",
    "if \"group\" not in adata.obs:\n",
    "    raise ValueError(\"The dataset must have a 'group' column for classification!\")\n",
    "\n",
    "# **Split into train (80%) and test (20%)**\n",
    "train_indices, test_indices = train_test_split(\n",
    "    adata.obs_names,\n",
    "    test_size=0.2,\n",
    "    stratify=adata.obs[\"group\"],\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# Create separate train and test datasets\n",
    "train_dataset = adata[train_indices, :].copy()\n",
    "test_dataset = adata[test_indices, :].copy()\n",
    "\n",
    "# Save them separately for direct use in training\n",
    "train_dataset.write_h5ad(\"train_finaldata.h5ad\")\n",
    "test_dataset.write_h5ad(\"test_finaldata.h5ad\")\n",
    "\n",
    "print(f\" Train dataset: {train_dataset.shape[0]} cells\")\n",
    "print(f\" Test dataset: {test_dataset.shape[0]} cells\")\n",
    "\n",
    "# **Load preprocessed train and test datasets**\n",
    "train_dataset = ad.read_h5ad(\"train_finaldata.h5ad\")\n",
    "test_dataset = ad.read_h5ad(\"test_finaldata.h5ad\")\n",
    "\n",
    "print(f\"Loaded train dataset: {train_dataset.shape}\")\n",
    "print(f\"Loaded test dataset: {test_dataset.shape}\")\n",
    "\n",
    "# **Convert UMI counts to integers**\n",
    "train_dataset.X = np.round(train_dataset.X).astype(np.int32)\n",
    "test_dataset.X = np.round(test_dataset.X).astype(np.int32)\n",
    "print(\"Converted gene expression matrix to integer UMI counts.\")\n",
    "\n",
    "# **Extract unique group labels and count**\n",
    "label_encoder = LabelEncoder()\n",
    "group_labels = sorted(train_dataset.obs[\"group\"].unique())\n",
    "num_groups = len(group_labels)\n",
    "print(f\"Found {num_groups} unique group labels.\")\n",
    "\n",
    "# **Configure scGPT**\n",
    "scgpt_config = scGPTConfig(batch_size=128, device=device)\n",
    "\n",
    "# **Initialize the fine-tuning model**\n",
    "scgpt_fine_tune = scGPTFineTuningModel(\n",
    "    scGPT_config=scgpt_config,\n",
    "    fine_tuning_head=\"classification\",\n",
    "    output_size=num_groups\n",
    ")\n",
    "print(\"scGPT fine-tuning model initialized successfully.\")\n",
    "\n",
    "final_marker_dict = {\n",
    "\"endothelial cell_lung cancer\": ['VWF', 'EPAS1', 'PECAM1', 'CLDN5', 'CLEC14A', 'SPARCL1', 'RAMP2', 'HYAL2', 'AQP1', 'TM4SF1'],\n",
    "\"endothelial cell_breast cancer\": ['RAMP2', 'GNG11', 'EGFL7', 'PLVAP', 'SPARCL1', 'PECAM1', 'CRIP2', 'CD34', 'ADGRL4', 'IGFBP7'],\n",
    "\"endothelial cell_melanoma\": ['SHISA9', 'KCNQ1OT1', 'ANKRD20A9P', 'ORC4', 'EEF1G', 'VIM', 'CFLAR', 'TNFAIP8L1', 'HLA-B', 'RBMS2'],\n",
    "\"endothelial cell_liver cancer\": ['RAMP2', 'EGFL7', 'HSPG2', 'GNG11', 'PLVAP', 'ENG', 'PECAM1', 'TCF4', 'FKBP1A', 'CRIP2'],\n",
    "\"endothelial cell_uveal melanoma\": ['GNG11', 'RAMP2', 'CAVIN2', 'IFITM3', 'PECAM1', 'TM4SF1', 'RAMP3', 'CRIP2', 'ESAM', 'ENG'],\n",
    "\"endothelial cell_colorectal cancer\": ['RAMP2', 'GNG11', 'PECAM1', 'FKBP1A', 'IGFBP7', 'CRIP2', 'TM4SF1', 'HSPG2', 'VWF', 'SPARCL1'],\n",
    "\"endothelial cell_ovarian cancer\": ['SPARCL1', 'GNG11', 'RAMP2', 'VWF', 'IGFBP7', 'CLDN5', 'PECAM1', 'TM4SF1', 'HSPG2', 'CRIP2'],\n",
    "\"mononuclear phagocyte_lung cancer\": ['TYROBP', 'LYZ', 'HLA-DRA', 'FCER1G', 'HLA-DRB1', 'CD74', 'LAPTM5', 'SRGN', 'GLUL', 'CTSS'],\n",
    "\"mononuclear phagocyte_breast cancer\": ['TYROBP', 'FCER1G', 'GPX1', 'AIF1', 'SPI1', 'CD74', 'GABARAP', 'CD68', 'CST3', 'HLA-DRB1'],\n",
    "\"mononuclear phagocyte_melanoma\": ['HLA-B', 'SAT1', 'HLA-C', 'SRGN', 'B2M', 'HLA-DRA', 'TAPBP', 'LYZ', 'METTL21A', 'ACTB'],\n",
    "\"mononuclear phagocyte_liver cancer\": ['AIF1', 'HLA-DRA', 'TYROBP', 'LST1', 'HLA-DPB1', 'HLA-DRB1', 'CST3', 'CD74', 'HLA-DPA1', 'LYZ'],\n",
    "\"mononuclear phagocyte_uveal melanoma\": ['TYROBP', 'CST3', 'CD74', 'AIF1', 'FCER1G', 'FTL', 'HLA-DRA', 'HLA-DPA1', 'CD68', 'MS4A7'],\n",
    "\"mononuclear phagocyte_colorectal cancer\": ['TYROBP', 'FCER1G', 'AIF1', 'LYZ', 'CYBA', 'FTL', 'GPX1', 'HLA-DRA', 'SRGN', 'LST1'],\n",
    "\"mononuclear phagocyte_ovarian cancer\": ['TYROBP', 'AIF1', 'HLA-DRA', 'FCER1G', 'HLA-DPB1', 'HLA-DPA1', 'HLA-DRB1', 'CD74', 'FTL', 'SAT1'],\n",
    "\"B cell_lung cancer\": ['CD79A', 'CD74', 'HLA-DRA', 'CXCR4', 'TTC19', 'MS4A1', 'CD37', 'HLA-DRB1'],\n",
    "\"B cell_breast cancer\": ['CD79A', 'CD74', 'MS4A1', 'NCF1', 'CD37', 'HLA-DRB5', 'HLA-DQA1'],\n",
    "\"B cell_melanoma\": ['CD74', 'ORAI2', 'UGDH-AS1', 'KCNQ1OT1', 'CFLAR', 'SHISA9', 'ORC4', 'TMEM212', 'ASTN2', 'METTL21A'],\n",
    "\"B cell_liver cancer\": ['CD79A', 'RPS29', 'RPS10', 'IGKC', 'CD74', 'RPLP2', 'RPS27', 'MS4A1', 'HBB', 'RPL27A'],\n",
    "\"B cell_uveal melanoma\": ['IGHG1', 'CD79A', 'JCHAIN', 'HERPUD1', 'MZB1', 'SSR4'],\n",
    "\"B cell_colorectal cancer\": ['CD79A', 'MS4A1', 'IGKC', 'CD37', 'RPL21', 'CD52', 'RPS29', 'CD79B', 'HLA-DRA', 'RPLP2'],\n",
    "\"B cell_ovarian cancer\": ['IGKC', 'IGLC2', 'IGHG3', 'IGHG1', 'IGHG4', 'CD79A', 'IGLC3', 'IGHGP', 'HLA-DRA', 'MZB1'],\n",
    "\"malignant cell_lung cancer\": ['KRT19', 'KRT8', 'KRT18', 'PERP', 'SPINT2', 'EPCAM', 'MGST1', 'TXN', 'KRT7', 'TACSTD2'],\n",
    "\"malignant cell_breast cancer\": ['MIF', 'KRT19', 'PPDPF', 'COX6C', 'GABARAP', 'H3-3A', 'RPL30', 'TCEA1', 'RPL8'],\n",
    "\"malignant cell_melanoma\": ['GAPDH', 'HSP90AB1', 'KCNQ1OT1', 'RPL17', 'RPS17', 'ACTG1', 'EEF1G', 'ALDOA', 'EEF1A1', 'HSP90AA1'],\n",
    "\"malignant cell_liver cancer\": ['KRT18', 'UQCRQ', 'SEC61G', 'KRT8', 'SERPINA1', 'ATP5MF', 'TM4SF4', 'ELOB', 'COX6A1'],\n",
    "\"malignant cell_colorectal cancer\": ['PHGR1', 'KRT8', 'S100A6', 'TFF3', 'LGALS4', 'KRT18', 'TSPAN8', 'EPCAM', 'KRT19', 'C19orf33'],\n",
    "\"malignant cell_ovarian cancer\": ['KRT18', 'KRT8', 'CD24', 'WFDC2', 'RPL7', 'GSTP1', 'MDK', 'KRT19', 'YWHAE', 'RBP1'],\n",
    "\"malignant cell_uveal melanoma\": ['S100A1', 'CITED1', 'SDCBP', 'MLANA', 'ZNF106', 'RAB38', 'MIF', 'EFHD1', 'ANXA5', 'ST3GAL4'],\n",
    "\"epithelial cell_lung cancer\": ['SFTPB', 'SFTPA2', 'NAPSA', 'SFTPA1', 'KRT19', 'SFTA2', 'HOPX', 'SCGB3A2'],\n",
    "\"epithelial cell_breast cancer\": ['MUCL1', 'SCGB2A2', 'IGKV3-20', 'IGHV3-23', 'IGKV4-1', 'IGLV1-51', 'KRT19', 'MGP', 'AZGP1', 'IGHG1'],\n",
    "\"epithelial cell_melanoma\": ['KCNQ1OT1', 'ANKRD20A9P', 'TOR1AIP2', 'MBOAT1', 'CFLAR', 'UGDH-AS1', 'CHP1', 'GAPDH', 'ORAI2', 'GATD1'],\n",
    "\"epithelial cell_liver cancer\": ['ALB', 'ATP1B1', 'SPP1', 'VTN', 'FXYD2', 'TM4SF4', 'PDZK1IP1', 'ANXA4', 'CLDN10', 'SERPINA1'],\n",
    "\"epithelial cell_colorectal cancer\": ['PHGR1', 'TFF3', 'LGALS4', 'KRT8', 'S100A6', 'KRT18', 'FXYD3', 'TSPAN8', 'S100P', 'AGR2'],\n",
    "\"epithelial cell_ovarian cancer\": ['HBB', 'HBA2', 'HBA1', 'WFDC2', 'COL1A1', 'COL1A2', 'TAGLN', 'COL3A1', 'RBP1', 'FABP4'],\n",
    "\"epithelial cell_uveal melanoma\": ['MLANA', 'TYRP1', 'PMEL', 'S100A1', 'SNHG7', 'MITF', 'TEX41', 'SLCO4A1-AS1', 'RAB38'],\n",
    "\"T cell_lung cancer\": ['CXCR4', 'CD3D', 'TRBC2', 'CD2', 'SRGN', 'RGS1', 'IL32', 'TRBC1', 'PTPRC', 'CD7'],\n",
    "\"T cell_breast cancer\": ['CD3E', 'IFITM1', 'RAC2', 'CORO1A', 'PTPRC', 'CD52', 'IL32'],\n",
    "\"T cell_melanoma\": ['ASTN2', 'UGDH-AS1', 'SHISA9', 'KCNQ1OT1', 'IKZF3', 'ORC4', 'CFLAR', 'TMEM212', 'ANKRD20A9P', 'B2M'],\n",
    "\"T cell_liver cancer\": ['RPS29', 'LTB', 'RPS27', 'RPLP2', 'CD3D', 'RPS10', 'CD52', 'RPL21', 'HBB', 'RPL39'],\n",
    "\"T cell_uveal melanoma\": ['CST7', 'CD3E', 'CREM', 'CD3D', 'PTPRC', 'SRGN', 'IL32', 'TNFAIP3', 'NKG7'],\n",
    "\"T cell_colorectal cancer\": ['CD3D', 'CD2', 'IL32', 'CD52', 'CORO1A', 'TMSB4X', 'RPS29', 'BTG1', 'MALAT1', 'CD3E'],\n",
    "\"T cell_ovarian cancer\": ['CXCR4', 'IL32', 'BTG1', 'CD2', 'CD3D', 'CD52', 'CORO1A', 'GZMA', 'IGKC', 'TRBC1'],\n",
    "\"mast cell_lung cancer\": ['CPA3', 'TPSAB1', 'TPSB2', 'MS4A2', 'KIT', 'SLC18A2', 'GATA2', 'HPGDS', 'VWA5A', 'SRGN'],\n",
    "\"mast cell_breast cancer\": ['TPSAB1', 'CPA3', 'TPSB2', 'MS4A2', 'GATA2', 'HPGDS', 'VWA5A', 'SLC18A2', 'LTC4S', 'HDC'],\n",
    "\"mast cell_melanoma\": ['TPSD1', 'CPA3', 'TPSAB1', 'TPSB2', 'MS4A2', 'HPGDS', 'CTSG', 'SLC18A2', 'VWA5A'],\n",
    "\"mast cell_liver cancer\": ['TPSAB1', 'TPSB2', 'HPGDS', 'IL1RL1', 'CPA3', 'LINC00623', 'REST', 'ALOX5AP', 'SRGN', 'ID2'],\n",
    "\"mast cell_colorectal cancer\": ['TPSAB1', 'CPA3', 'TPSB2', 'MS4A2', 'HPGDS', 'SAMSN1', 'LTC4S', 'HPGD', 'RGS1', 'ALOX5AP'],\n",
    "\"mast cell_ovarian cancer\": ['TPSAB1', 'TPSB2', 'CPA3', 'HPGDS', 'MS4A2', 'RGS13', 'VWA5A', 'FCER1G', 'ALOX5AP', 'LTC4S'],\n",
    "\"fibroblast_lung cancer\": ['DCN', 'COL1A2', 'SPARC', 'COL6A2', 'BGN', 'COL3A1', 'SERPING1', 'C1R', 'LUM', 'C1S'],\n",
    "\"fibroblast_breast cancer\": ['COL1A2', 'COL1A1', 'COL3A1', 'SPARC', 'COL6A2', 'COL6A1', 'CALD1', 'COL6A3', 'BGN', 'AEBP1'],\n",
    "\"fibroblast_melanoma\": ['EEF1G', 'KCNQ1OT1', 'UGDH-AS1', 'VIM', 'CHP1', 'RPL13AP5', 'ACTB', 'EEF1A1', 'RPS17', 'ACTG1'],\n",
    "\"fibroblast_liver cancer\": ['IGFBP7', 'CALD1', 'BGN', 'MYL9', 'RGS5', 'PPP1R14A', 'MGP', 'TAGLN', 'PLAC9', 'SPARC'],\n",
    "\"fibroblast_uveal melanoma\": ['FRZB', 'CST3', 'TIMP3', 'IGFBP7', 'SELENOW', 'RBP1', 'SERPINF1', 'PTGDS'],\n",
    "\"fibroblast_colorectal cancer\": ['CALD1', 'LGALS1', 'TPM2', 'IGFBP7', 'SELENOM', 'COL3A1', 'SPARC', 'TIMP1', 'COL1A2', 'COL6A2'],\n",
    "\"fibroblast_ovarian cancer\": ['COL1A2', 'COL1A1', 'DCN', 'CALD1', 'COL3A1', 'LUM', 'SPARC', 'LGALS1', 'RARRES2', 'TPM2'],\n",
    "\"plasmacytoid dendritic cell_lung cancer\": ['GZMB', 'GPR183', 'JCHAIN', 'PLAC8', 'IRF7', 'IRF4', 'CD74', 'RGS2', 'TCF4', 'PLD4'],\n",
    "\"plasmacytoid dendritic cell_breast cancer\": ['JCHAIN', 'IRF8', 'IRF7', 'LILRA4', 'PLAC8', 'GZMB', 'ITM2C', 'PLD4', 'CD74', 'CCDC50'],\n",
    "\"plasmacytoid dendritic cell_melanoma\": ['IRF7', 'MPEG1', 'PLD4', 'PLEK', 'IRF8', 'JCHAIN', 'PLAC8', 'GZMB'],\n",
    "\"plasmacytoid dendritic cell_liver cancer\": ['JCHAIN', 'IRF7', 'GZMB', 'PLAC8', 'GPR183', 'PPP1R14B', 'IRF8', 'RGS2', 'ITM2C', 'SEC61B'],\n",
    "\"plasmacytoid dendritic cell_uveal melanoma\": ['GZMB', 'LILRA4', 'JCHAIN', 'PLD4', 'IRF8', 'IRF7', 'SEC61B', 'CD74', 'PLAC8', 'GPR183'],\n",
    "\"plasmacytoid dendritic cell_colorectal cancer\": ['JCHAIN', 'GZMB', 'IRF8', 'PLAC8', 'IRF7', 'SEC61B', 'UGCG', 'IRF4', 'CCDC50', 'CLN8'],\n",
    "\"plasmacytoid dendritic cell_ovarian cancer\": ['GZMB', 'JCHAIN', 'GPR183', 'IRF7', 'PLAC8', 'ALOX5AP', 'AREG', 'C12orf75', 'TSPAN13', 'CD74'],\n",
    "\"neutrophil_lung cancer\": ['G0S2', 'CXCL8', 'NAMPT', 'ALOX5AP', 'SRGN', 'BCL2A1', 'PLAUR', 'LITAF'],\n",
    "\"neutrophil_uveal melanoma\": ['FCGR3B', 'CSF3R', 'S100A8', 'MNDA', 'S100A9', 'NAMPT', 'RGS2', 'FPR1', 'RIPOR2', 'MXD1'],\n",
    "\"neutrophil_colorectal cancer\": ['CXCL8', 'BCL2A1', 'S100A8', 'NAMPT', 'G0S2', 'S100A9', 'SAT1', 'FTH1', 'ALOX5AP', 'PLEK'],\n",
    "\"neutrophil_ovarian cancer\": ['CXCL8', 'BCL2A1', 'C15orf48', 'PELATON', 'S100A9', 'G0S2', 'NAMPT', 'FTH1', 'MXD1', 'S100A8']\n",
    "}\n",
    "\n",
    "# **(Optional) Set marker genes if using attention supervision**\n",
    "scgpt_fine_tune.model.marker_gene_dict = final_marker_dict  # Define this earlier if needed\n",
    "scgpt_fine_tune.model.lambda_attn = 0.1  # Tune this weight\n",
    "\n",
    "# **Process datasets**\n",
    "dataset = scgpt_fine_tune.process_data(train_dataset, gene_names=\"feature_name\")\n",
    "validation_dataset = scgpt_fine_tune.process_data(test_dataset, gene_names=\"feature_name\")\n",
    "\n",
    "# **Check dataset sizes**\n",
    "print(f\"Processed training dataset size: {len(dataset)}\")\n",
    "print(f\"Processed validation dataset size: {len(validation_dataset)}\")\n",
    "\n",
    "if len(dataset) == 0 or len(validation_dataset) == 0:\n",
    "    raise ValueError(\"Processed dataset is empty! Check gene names or dataset formatting.\")\n",
    "\n",
    "# **Encode group labels**\n",
    "label_encoder.fit(group_labels)\n",
    "group_labels_train = torch.tensor(label_encoder.transform(train_dataset.obs[\"group\"]), dtype=torch.long)\n",
    "group_labels_test = torch.tensor(label_encoder.transform(test_dataset.obs[\"group\"]), dtype=torch.long)\n",
    "\n",
    "print(f\"Encoded {len(label_encoder.classes_)} unique groups.\")\n",
    "\n",
    "# **Compute Raw Class Weights**\n",
    "class_counts = train_dataset.obs[\"group\"].value_counts().to_dict()\n",
    "total_samples = sum(class_counts.values())\n",
    "\n",
    "# **Ensure class alignment**\n",
    "raw_weights = torch.tensor(\n",
    "    [total_samples / class_counts.get(cls, 1) for cls in group_labels], dtype=torch.float32\n",
    ")\n",
    "\n",
    "# **Apply Log Scaling for Stability**\n",
    "log_weights = torch.log1p(raw_weights)\n",
    "class_weights = log_weights / log_weights.mean()\n",
    "class_weights = class_weights.to(device)\n",
    "\n",
    "print(\"Log-scaled normalized class weights:\", class_weights)\n",
    "\n",
    "# **Apply Class Weights to CrossEntropyLoss**\n",
    "loss_fn = CrossEntropyLoss(weight=class_weights)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b5eec79e-319c-44c1-bf0b-0bbbe2effb32",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helical.models.scgpt.fine_tuning_model:Starting Fine-Tuning\n",
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/helical/models/scgpt/fine_tuning_model.py:231: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  validation_batch_count = 0\n",
      "\n",
      "Fine-Tuning: epoch 1/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=4.33]\n",
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/helical/models/scgpt/fine_tuning_model.py:268: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  The dataset to get the outputs from.\n",
      "\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=4.33]\n",
      "Fine-Tuning: epoch 2/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=3.77]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=3.27]\n",
      "Fine-Tuning: epoch 3/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=2.67]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=2.24]\n",
      "Fine-Tuning: epoch 4/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=1.88]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.84it/s, val_loss=1.61]\n",
      "Fine-Tuning: epoch 5/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=1.38]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=1.19]\n",
      "Fine-Tuning: epoch 6/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=1.03]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.896]\n",
      "Fine-Tuning: epoch 7/25: 100%|███████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.788]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.697]\n",
      "Fine-Tuning: epoch 8/25: 100%|███████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.629]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.573]\n",
      "Fine-Tuning: epoch 9/25: 100%|███████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.524]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.84it/s, val_loss=0.489]\n",
      "Fine-Tuning: epoch 10/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.453]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.433]\n",
      "Fine-Tuning: epoch 11/25: 100%|████████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.4]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.394]\n",
      "Fine-Tuning: epoch 12/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.362]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.361]\n",
      "Fine-Tuning: epoch 13/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.334]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.336]\n",
      "Fine-Tuning: epoch 14/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.311]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.317]\n",
      "Fine-Tuning: epoch 15/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.291]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.303]\n",
      "Fine-Tuning: epoch 16/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.275]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.89it/s, val_loss=0.292]\n",
      "Fine-Tuning: epoch 17/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.264]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.279]\n",
      "Fine-Tuning: epoch 18/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.252]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.27]\n",
      "Fine-Tuning: epoch 19/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.244]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.264]\n",
      "Fine-Tuning: epoch 20/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.234]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.257]\n",
      "Fine-Tuning: epoch 21/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.225]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.25]\n",
      "Fine-Tuning: epoch 22/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.219]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.249]\n",
      "Fine-Tuning: epoch 23/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.212]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.245]\n",
      "Fine-Tuning: epoch 24/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:27<00:00,  3.79it/s, loss=0.206]\n",
      "Fine-Tuning Validation: 100%|███████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.24]\n",
      "Fine-Tuning: epoch 25/25: 100%|██████████████████████████████████████████████████████████████████| 1467/1467 [06:26<00:00,  3.79it/s, loss=0.201]\n",
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████| 367/367 [00:37<00:00,  9.90it/s, val_loss=0.236]\n",
      "INFO:helical.models.scgpt.fine_tuning_model:Fine-Tuning Complete. Epochs: 25\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training completed successfully!\n",
      "Fine-tuned model saved as 'scgpt_final_marker_attn.pth'.\n"
     ]
    }
   ],
   "source": [
    "scgpt_fine_tune.train(\n",
    "    train_input_data=dataset,\n",
    "    train_labels=group_labels_train,\n",
    "    validation_input_data=validation_dataset,\n",
    "    validation_labels=group_labels_test,\n",
    "    epochs=25,\n",
    "    optimizer_params={\"lr\": 2e-5, \"weight_decay\": 1e-4},\n",
    "    lr_scheduler_params={\n",
    "        \"name\": \"linear\",\n",
    "        \"num_warmup_steps\": 100,\n",
    "        \"num_training_steps\": len(dataset) * 10\n",
    "    },\n",
    "    loss_function=loss_fn\n",
    ")\n",
    "\n",
    "print(\"Training completed successfully!\")\n",
    "\n",
    "# ✅ Save the fine-tuned model\n",
    "torch.save(scgpt_fine_tune.state_dict(), \"scgpt_markerattn.pth\")\n",
    "print(\"Fine-tuned model saved as 'scgpt_final_marker_attn.pth'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0c488cf7-e7e7-4cfc-ab34-2e8e73b7f351",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group labels: 66 unique classes\n"
     ]
    }
   ],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "import torch\n",
    "\n",
    "# Ensure 'group' is categorical (group = cell type + cancer type)\n",
    "train_labels = train_dataset.obs[\"group\"].astype(\"category\")\n",
    "test_labels = test_dataset.obs[\"group\"].astype(\"category\")\n",
    "\n",
    "# Encode group labels as numerical labels\n",
    "label_encoder = LabelEncoder()\n",
    "group_labels_train = label_encoder.fit_transform(train_labels)\n",
    "group_labels_test = label_encoder.transform(test_labels)\n",
    "\n",
    "# Convert to PyTorch tensor (scGPT expects tensors)\n",
    "group_labels_train = torch.tensor(group_labels_train, dtype=torch.long)\n",
    "group_labels_test = torch.tensor(group_labels_test, dtype=torch.long)\n",
    "\n",
    "print(f\"Group labels: {len(label_encoder.classes_)} unique classes\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f7cab386-8c38-4912-a59a-ab04019cef4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████████████████████| 367/367 [00:35<00:00, 10.43it/s]\n"
     ]
    }
   ],
   "source": [
    "outputs = scgpt_fine_tune.get_outputs(validation_dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3fbf50e2-c0a4-4b99-a2c2-a64d8ca08b70",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:helical.models.scgpt.model:Started getting embeddings:\n",
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/helical/models/scgpt/model.py:142: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.\n",
      "  torch.cuda.amp.autocast(enabled=True),\n",
      "\n",
      "Embedding cells: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 367/367 [00:16<00:00, 22.05it/s]\n",
      "INFO:helical.models.scgpt.model:Finished getting embeddings.\n"
     ]
    }
   ],
   "source": [
    "embeddings = scgpt_fine_tune.get_embeddings(validation_dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "32e3db02-e108-420c-a9cc-4ce9003c32dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:py.warnings:/home/aigpu/miniconda3/envs/OmicsFM/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Accuracy': 0.9240654703550574,\n",
       " 'Precision': 0.9251111742860351,\n",
       " 'Recall': 0.9240654703550574,\n",
       " 'F1 Score': 0.9240700466694403}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "\n",
    "# Convert predictions and true labels to NumPy arrays (if not already)\n",
    "y_true = group_labels_test.numpy() if isinstance(group_labels_test, torch.Tensor) else group_labels_test\n",
    "y_pred = outputs.argmax(dim=1).cpu().numpy() if isinstance(outputs, torch.Tensor) else outputs.argmax(axis=1)\n",
    "\n",
    "# Compute evaluation metrics\n",
    "accuracy = accuracy_score(y_true, y_pred)\n",
    "precision = precision_score(y_true, y_pred, average='weighted')\n",
    "recall = recall_score(y_true, y_pred, average='weighted')\n",
    "f1 = f1_score(y_true, y_pred, average='weighted')\n",
    "\n",
    "# Display results\n",
    "metrics = {\n",
    "    \"Accuracy\": accuracy,\n",
    "    \"Precision\": precision,\n",
    "    \"Recall\": recall,\n",
    "    \"F1 Score\": f1\n",
    "}\n",
    "metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e4821eb8-1dc5-4129-a5ce-7e8147e56fdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>True Group</th>\n",
       "      <th>Predicted Group</th>\n",
       "      <th>Top Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[CYBA, HERPUD1, MGP, PNLIPRP3, ENSG00000224500...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>malignant cell_lung cancer</td>\n",
       "      <td>malignant cell_lung cancer</td>\n",
       "      <td>[CLEC3A, COL1A1, CRABP1, MCEMP1, HERPUD1, MACR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fibroblast_colorectal cancer</td>\n",
       "      <td>fibroblast_colorectal cancer</td>\n",
       "      <td>[GZMK, CHRNA4, ENSG00000237993.1, CLEC3A, HTR5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>[CLEC3A, AIF1, ENSG00000237993.1, EPCAM, RPL8,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[CXCL13, C4orf50, COX6A1, IFNG, PNLIPRP3, SAMS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>[RUBCNL, ACTB, IFNG, CD74, ENSG00000237993.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[ENSG00000224500.1, CLEC3A, IL32, IL1RN, SAMSN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>endothelial cell_ovarian cancer</td>\n",
       "      <td>endothelial cell_ovarian cancer</td>\n",
       "      <td>[MACROD2-IT1, COL6A3, MIR28, ENSG00000237993.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>malignant cell_uveal melanoma</td>\n",
       "      <td>malignant cell_uveal melanoma</td>\n",
       "      <td>[ENSG00000176593.8, PAK6-AS1, VCX2, TAAR8, GNG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>B cell_ovarian cancer</td>\n",
       "      <td>[ENSG00000237993.1, IFNG, SAMSN1, CXCL13, FTL,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mononuclear phagocyte_colorectal cancer</td>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>[PNLIPRP3, LELP1, ENSG00000203620.2, LINC00970...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>[ACTB, LINC00970, VCX2, PNLIPRP3, PTPRC, CLEC3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>[CYBA, CXCR6, PLAC8, SFTPA1, TEX33, BIRC3, CD7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[TRAV40, CD74, RNU1-82P, MCEMP1, RUBCNL, FCN2,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mononuclear phagocyte_uveal melanoma</td>\n",
       "      <td>mononuclear phagocyte_uveal melanoma</td>\n",
       "      <td>[FTL, ENSG00000237993.1, MYL9, MAST4-IT1, SCGB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>[ENSG00000176593.8, TAAR8, TMIGD2, RNU1-82P, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>epithelial cell_colorectal cancer</td>\n",
       "      <td>epithelial cell_colorectal cancer</td>\n",
       "      <td>[RNU1-82P, CLEC3A, CRABP1, TMIGD2, OR52K1, ENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[ENSG00000237993.1, AZGP1, PAK6-AS1, ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>[CLDN10, VIM, MYL9, CD74, TYROBP, PLAUR, AIF1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[TRAV38-1, TRAV27, MIR28, ENSG00000237993.1, K...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 True Group  \\\n",
       "0                        T cell_lung cancer   \n",
       "1                malignant cell_lung cancer   \n",
       "2              fibroblast_colorectal cancer   \n",
       "3                     T cell_uveal melanoma   \n",
       "4                        T cell_lung cancer   \n",
       "5                           B cell_melanoma   \n",
       "6                        T cell_lung cancer   \n",
       "7           endothelial cell_ovarian cancer   \n",
       "8             malignant cell_uveal melanoma   \n",
       "9             malignant cell_ovarian cancer   \n",
       "10  mononuclear phagocyte_colorectal cancer   \n",
       "11                          B cell_melanoma   \n",
       "12                    T cell_uveal melanoma   \n",
       "13                       B cell_lung cancer   \n",
       "14     mononuclear phagocyte_uveal melanoma   \n",
       "15            malignant cell_ovarian cancer   \n",
       "16        epithelial cell_colorectal cancer   \n",
       "17                       B cell_lung cancer   \n",
       "18     mononuclear phagocyte_ovarian cancer   \n",
       "19                       B cell_lung cancer   \n",
       "\n",
       "                         Predicted Group  \\\n",
       "0                     T cell_lung cancer   \n",
       "1             malignant cell_lung cancer   \n",
       "2           fibroblast_colorectal cancer   \n",
       "3                  T cell_uveal melanoma   \n",
       "4                     T cell_lung cancer   \n",
       "5                        B cell_melanoma   \n",
       "6                     T cell_lung cancer   \n",
       "7        endothelial cell_ovarian cancer   \n",
       "8          malignant cell_uveal melanoma   \n",
       "9                  B cell_ovarian cancer   \n",
       "10  mononuclear phagocyte_ovarian cancer   \n",
       "11                       B cell_melanoma   \n",
       "12                 T cell_uveal melanoma   \n",
       "13                    B cell_lung cancer   \n",
       "14  mononuclear phagocyte_uveal melanoma   \n",
       "15         malignant cell_ovarian cancer   \n",
       "16     epithelial cell_colorectal cancer   \n",
       "17                    B cell_lung cancer   \n",
       "18  mononuclear phagocyte_ovarian cancer   \n",
       "19                    B cell_lung cancer   \n",
       "\n",
       "                                            Top Genes  \n",
       "0   [CYBA, HERPUD1, MGP, PNLIPRP3, ENSG00000224500...  \n",
       "1   [CLEC3A, COL1A1, CRABP1, MCEMP1, HERPUD1, MACR...  \n",
       "2   [GZMK, CHRNA4, ENSG00000237993.1, CLEC3A, HTR5...  \n",
       "3   [CLEC3A, AIF1, ENSG00000237993.1, EPCAM, RPL8,...  \n",
       "4   [CXCL13, C4orf50, COX6A1, IFNG, PNLIPRP3, SAMS...  \n",
       "5   [RUBCNL, ACTB, IFNG, CD74, ENSG00000237993.1, ...  \n",
       "6   [ENSG00000224500.1, CLEC3A, IL32, IL1RN, SAMSN...  \n",
       "7   [MACROD2-IT1, COL6A3, MIR28, ENSG00000237993.1...  \n",
       "8   [ENSG00000176593.8, PAK6-AS1, VCX2, TAAR8, GNG...  \n",
       "9   [ENSG00000237993.1, IFNG, SAMSN1, CXCL13, FTL,...  \n",
       "10  [PNLIPRP3, LELP1, ENSG00000203620.2, LINC00970...  \n",
       "11  [ACTB, LINC00970, VCX2, PNLIPRP3, PTPRC, CLEC3...  \n",
       "12  [CYBA, CXCR6, PLAC8, SFTPA1, TEX33, BIRC3, CD7...  \n",
       "13  [TRAV40, CD74, RNU1-82P, MCEMP1, RUBCNL, FCN2,...  \n",
       "14  [FTL, ENSG00000237993.1, MYL9, MAST4-IT1, SCGB...  \n",
       "15  [ENSG00000176593.8, TAAR8, TMIGD2, RNU1-82P, M...  \n",
       "16  [RNU1-82P, CLEC3A, CRABP1, TMIGD2, OR52K1, ENS...  \n",
       "17  [ENSG00000237993.1, AZGP1, PAK6-AS1, ENSG00000...  \n",
       "18  [CLDN10, VIM, MYL9, CD74, TYROBP, PLAUR, AIF1,...  \n",
       "19  [TRAV38-1, TRAV27, MIR28, ENSG00000237993.1, K...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import torch\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "# Step 1: Prepare validation data\n",
    "test_data = validation_dataset.count_matrix\n",
    "if issparse(test_data):\n",
    "    test_data = test_data.toarray()\n",
    "test_data = test_data.astype(int)\n",
    "\n",
    "gene_names = adata.var[\"feature_name\"].tolist()\n",
    "num_samples, seq_len = test_data.shape\n",
    "batch_size = 32\n",
    "\n",
    "# Step 2: Define function to get attention-based importance per sample\n",
    "def get_attention_weights_batch(model, input_tensor):\n",
    "    with torch.no_grad():\n",
    "        embedded_input = model.encoder(input_tensor)\n",
    "        attn_list = []\n",
    "        for layer in model.transformer_encoder.layers:\n",
    "            _, attn_weight = layer.self_attn(\n",
    "                embedded_input, embedded_input, embedded_input, need_weights=True\n",
    "            )\n",
    "            attn_list.append(attn_weight)\n",
    "        attn_stack = torch.stack(attn_list)  # [num_layers, B, seq_len, seq_len]\n",
    "        attn_avg = attn_stack.mean(dim=0)    # [B, seq_len, seq_len]\n",
    "        gene_importance = attn_avg.sum(dim=1)  # Sum over source: [B, seq_len]\n",
    "    return gene_importance\n",
    "\n",
    "# Step 3: Run attention scoring in batches\n",
    "all_gene_importance = []\n",
    "for i in range(0, num_samples, batch_size):\n",
    "    batch = test_data[i:i + batch_size]\n",
    "    batch_tensor = torch.tensor(batch, dtype=torch.long).to(device)\n",
    "    gene_imp_batch = get_attention_weights_batch(scgpt_fine_tune.model, batch_tensor)\n",
    "    all_gene_importance.append(gene_imp_batch.cpu().numpy())\n",
    "\n",
    "all_gene_importance = np.concatenate(all_gene_importance, axis=0)  # [N, seq_len]\n",
    "\n",
    "# Step 4: Map label indices to group names\n",
    "id_class_dict = {i: label for i, label in enumerate(label_encoder.classes_)}\n",
    "true_groups = [id_class_dict[label] for label in y_true]\n",
    "predicted_groups = [id_class_dict[label] for label in y_pred]\n",
    "\n",
    "# Step 5: For each sample, extract top attended genes\n",
    "results = []\n",
    "for i in range(num_samples):\n",
    "    gene_imp = all_gene_importance[i]\n",
    "    top_indices = np.argsort(gene_imp)[::-1][:10]\n",
    "    top_genes = [gene_names[idx] for idx in top_indices]\n",
    "    results.append({\n",
    "        \"True Group\": true_groups[i],\n",
    "        \"Predicted Group\": predicted_groups[i],\n",
    "        \"Top Genes\": top_genes\n",
    "    })\n",
    "\n",
    "final_df = pd.DataFrame(results)\n",
    "display(final_df.head(20))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "7076dcc1-099a-4e3b-952a-4313d2c25a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "tokens = input_ids_list[i]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6e79a714-349f-4760-8c26-90407d245163",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch attention shape: torch.Size([10, 123])\n"
     ]
    }
   ],
   "source": [
    "print(\"Batch attention shape:\", attn_scores.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "07c63e0b-95f3-4895-8fe7-22c116a6af3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: max_length used in collator = 1200\n"
     ]
    }
   ],
   "source": [
    "print(\"DEBUG: max_length used in collator =\", collator.max_length)\n",
    "\n",
    "val_loader = DataLoader(\n",
    "    validation_dataset,\n",
    "    batch_size=32,\n",
    "    sampler=SequentialSampler(validation_dataset),\n",
    "    collate_fn=collator,\n",
    "    drop_last=False,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6f53ab8c-d2ce-48fa-b6f8-c89069de8d16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch 0 input shape: torch.Size([32, 168])\n"
     ]
    }
   ],
   "source": [
    "for i, batch in enumerate(val_loader):\n",
    "    input_ids = batch[\"gene\"].to(device)\n",
    "    print(f\"Batch {i} input shape: {input_ids.shape}\")  # This should be [B, 1200]\n",
    "    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "341491bd-6a8f-450d-9340-2b3f41b08fc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch 0 - attn shape: torch.Size([10, 123])\n"
     ]
    }
   ],
   "source": [
    "print(f\"Batch {i} - attn shape: {attn_scores.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "7a427b05-5a32-4855-8f80-58ddd73acc1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████████████████████| 367/367 [00:35<00:00, 10.43it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>True Group</th>\n",
       "      <th>Predicted Group</th>\n",
       "      <th>Top Attended Marker Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[CD2, TRBC2, TRBC1, CD3D, CD7, RGS1, IL32, CXC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>malignant cell_lung cancer</td>\n",
       "      <td>malignant cell_lung cancer</td>\n",
       "      <td>[EPCAM, KRT19, SPINT2, KRT18, MGST1, TXN]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fibroblast_colorectal cancer</td>\n",
       "      <td>fibroblast_colorectal cancer</td>\n",
       "      <td>[SPARC, COL6A2, COL1A2, TPM2, COL3A1, LGALS1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>[CST7, CD3D, CREM, CD3E, IL32, NKG7, SRGN, TNF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[TRBC2, CD3D, RGS1, IL32, PTPRC, CXCR4, SRGN]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>[TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[CD2, CD7, CD3D, RGS1, IL32, CXCR4, SRGN]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>endothelial cell_ovarian cancer</td>\n",
       "      <td>endothelial cell_ovarian cancer</td>\n",
       "      <td>[RAMP2, CLDN5, HSPG2, VWF, CRIP2, GNG11, PECAM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>malignant cell_uveal melanoma</td>\n",
       "      <td>malignant cell_uveal melanoma</td>\n",
       "      <td>[EFHD1, ST3GAL4, S100A1, MLANA, CITED1, RAB38,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>B cell_ovarian cancer</td>\n",
       "      <td>[MDK, CD24, RPL7, WFDC2, GSTP1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mononuclear phagocyte_colorectal cancer</td>\n",
       "      <td>mononuclear phagocyte_colorectal cancer</td>\n",
       "      <td>[TYROBP, FTL, FCER1G, LYZ, AIF1, LST1, HLA-DRA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>[TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>[CREM, CST7, CD3D, CD3E, IL32, SRGN, PTPRC, TN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[CD74]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mononuclear phagocyte_uveal melanoma</td>\n",
       "      <td>mononuclear phagocyte_uveal melanoma</td>\n",
       "      <td>[MS4A7, FTL, TYROBP, CST3, CD74]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>[KRT8, CD24, RPL7, KRT18, KRT19, MDK, WFDC2, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>epithelial cell_colorectal cancer</td>\n",
       "      <td>epithelial cell_colorectal cancer</td>\n",
       "      <td>[PHGR1, S100P, LGALS4, FXYD3, KRT8, TSPAN8, TF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[CD79A, CD37, HLA-DRB1, MS4A1, HLA-DRA, CXCR4,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>[HLA-DPB1, TYROBP, FTL, FCER1G, HLA-DRB1, AIF1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[TTC19]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 True Group  \\\n",
       "0                        T cell_lung cancer   \n",
       "1                malignant cell_lung cancer   \n",
       "2              fibroblast_colorectal cancer   \n",
       "3                     T cell_uveal melanoma   \n",
       "4                        T cell_lung cancer   \n",
       "5                           B cell_melanoma   \n",
       "6                        T cell_lung cancer   \n",
       "7           endothelial cell_ovarian cancer   \n",
       "8             malignant cell_uveal melanoma   \n",
       "9             malignant cell_ovarian cancer   \n",
       "10  mononuclear phagocyte_colorectal cancer   \n",
       "11                          B cell_melanoma   \n",
       "12                    T cell_uveal melanoma   \n",
       "13                       B cell_lung cancer   \n",
       "14     mononuclear phagocyte_uveal melanoma   \n",
       "15            malignant cell_ovarian cancer   \n",
       "16        epithelial cell_colorectal cancer   \n",
       "17                       B cell_lung cancer   \n",
       "18     mononuclear phagocyte_ovarian cancer   \n",
       "19                       B cell_lung cancer   \n",
       "\n",
       "                            Predicted Group  \\\n",
       "0                        T cell_lung cancer   \n",
       "1                malignant cell_lung cancer   \n",
       "2              fibroblast_colorectal cancer   \n",
       "3                     T cell_uveal melanoma   \n",
       "4                        T cell_lung cancer   \n",
       "5                           B cell_melanoma   \n",
       "6                        T cell_lung cancer   \n",
       "7           endothelial cell_ovarian cancer   \n",
       "8             malignant cell_uveal melanoma   \n",
       "9                     B cell_ovarian cancer   \n",
       "10  mononuclear phagocyte_colorectal cancer   \n",
       "11                          B cell_melanoma   \n",
       "12                    T cell_uveal melanoma   \n",
       "13                       B cell_lung cancer   \n",
       "14     mononuclear phagocyte_uveal melanoma   \n",
       "15            malignant cell_ovarian cancer   \n",
       "16        epithelial cell_colorectal cancer   \n",
       "17                       B cell_lung cancer   \n",
       "18     mononuclear phagocyte_ovarian cancer   \n",
       "19                       B cell_lung cancer   \n",
       "\n",
       "                            Top Attended Marker Genes  \n",
       "0   [CD2, TRBC2, TRBC1, CD3D, CD7, RGS1, IL32, CXC...  \n",
       "1           [EPCAM, KRT19, SPINT2, KRT18, MGST1, TXN]  \n",
       "2   [SPARC, COL6A2, COL1A2, TPM2, COL3A1, LGALS1, ...  \n",
       "3   [CST7, CD3D, CREM, CD3E, IL32, NKG7, SRGN, TNF...  \n",
       "4       [TRBC2, CD3D, RGS1, IL32, PTPRC, CXCR4, SRGN]  \n",
       "5   [TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4, ...  \n",
       "6           [CD2, CD7, CD3D, RGS1, IL32, CXCR4, SRGN]  \n",
       "7   [RAMP2, CLDN5, HSPG2, VWF, CRIP2, GNG11, PECAM...  \n",
       "8   [EFHD1, ST3GAL4, S100A1, MLANA, CITED1, RAB38,...  \n",
       "9                     [MDK, CD24, RPL7, WFDC2, GSTP1]  \n",
       "10  [TYROBP, FTL, FCER1G, LYZ, AIF1, LST1, HLA-DRA...  \n",
       "11  [TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4, ...  \n",
       "12  [CREM, CST7, CD3D, CD3E, IL32, SRGN, PTPRC, TN...  \n",
       "13                                             [CD74]  \n",
       "14                   [MS4A7, FTL, TYROBP, CST3, CD74]  \n",
       "15  [KRT8, CD24, RPL7, KRT18, KRT19, MDK, WFDC2, G...  \n",
       "16  [PHGR1, S100P, LGALS4, FXYD3, KRT8, TSPAN8, TF...  \n",
       "17  [CD79A, CD37, HLA-DRB1, MS4A1, HLA-DRA, CXCR4,...  \n",
       "18  [HLA-DPB1, TYROBP, FTL, FCER1G, HLA-DRB1, AIF1...  \n",
       "19                                            [TTC19]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import torch\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from torch.utils.data import DataLoader, SequentialSampler\n",
    "from helical.models.scgpt.data_collator import DataCollator\n",
    "\n",
    "# Ensure eval mode\n",
    "scgpt_fine_tune.model.eval()\n",
    "vocab = scgpt_fine_tune.vocab\n",
    "id_to_gene = {v: k for k, v in vocab.items()}\n",
    "\n",
    "# --- Collator with fixed padding ---\n",
    "collator = DataCollator(\n",
    "    do_padding=True,\n",
    "    pad_token_id=vocab[\"<pad>\"],\n",
    "    pad_value=0,\n",
    "    max_length=1200,\n",
    "    do_binning=True,\n",
    "    do_mlm=False,\n",
    "    sampling=False,\n",
    "    keep_first_n_tokens=1,\n",
    ")\n",
    "\n",
    "val_loader = DataLoader(\n",
    "    validation_dataset,\n",
    "    batch_size=32,\n",
    "    sampler=SequentialSampler(validation_dataset),\n",
    "    collate_fn=collator,\n",
    "    drop_last=False,\n",
    ")\n",
    "\n",
    "# --- Prediction outputs ---\n",
    "outputs = scgpt_fine_tune.get_outputs(validation_dataset)\n",
    "y_pred = np.argmax(outputs, axis=1)\n",
    "y_true = group_labels_test.numpy() if isinstance(group_labels_test, torch.Tensor) else group_labels_test\n",
    "id_class_dict = {i: c for i, c in enumerate(label_encoder.classes_)}\n",
    "true_groups = [id_class_dict[idx] for idx in y_true]\n",
    "pred_groups = [id_class_dict[idx] for idx in y_pred]\n",
    "\n",
    "# --- Map marker genes to token IDs ---\n",
    "group_to_marker_ids = {}\n",
    "for group, genes in final_marker_dict.items():\n",
    "    marker_ids = [vocab[g] for g in genes if g in vocab]\n",
    "    group_to_marker_ids[group] = marker_ids\n",
    "\n",
    "# --- CLS attention extractor ---\n",
    "def get_cls_attention(model, input_tensor):\n",
    "    with torch.no_grad():\n",
    "        embedded = model.encoder(input_tensor)\n",
    "        attn_list = []\n",
    "        for layer in model.transformer_encoder.layers:\n",
    "            attn = layer.self_attn(embedded, embedded, embedded, need_weights=True)[1]\n",
    "            cls_attn = attn[:, 0, :]  # CLS to each token\n",
    "            attn_list.append(cls_attn)\n",
    "        attn_avg = torch.stack(attn_list).mean(dim=0)  # (B, seq)\n",
    "    return attn_avg\n",
    "\n",
    "# --- Extract top attended marker genes per sample ---\n",
    "results = []\n",
    "sample_idx = 0\n",
    "\n",
    "for batch in val_loader:\n",
    "    device = next(scgpt_fine_tune.model.parameters()).device\n",
    "    input_ids = batch[\"gene\"].to(device)\n",
    "    attn_scores = get_cls_attention(scgpt_fine_tune.model, input_ids)  # (B, seq)\n",
    "\n",
    "    for i in range(attn_scores.shape[0]):\n",
    "        sample_input = input_ids[i].cpu().numpy()\n",
    "        sample_attn = attn_scores[i].cpu().numpy()\n",
    "        true_group = true_groups[sample_idx]\n",
    "        pred_group = pred_groups[sample_idx]\n",
    "\n",
    "        marker_ids = group_to_marker_ids.get(true_group, [])\n",
    "        top_marker_attn = [\n",
    "            (id_to_gene[token_id], sample_attn[j])\n",
    "            for j, token_id in enumerate(sample_input)\n",
    "            if token_id in marker_ids\n",
    "        ]\n",
    "        top_marker_attn.sort(key=lambda x: x[1], reverse=True)\n",
    "        top_genes = [g for g, _ in top_marker_attn[:10]]\n",
    "\n",
    "        results.append({\n",
    "            \"True Group\": true_group,\n",
    "            \"Predicted Group\": pred_group,\n",
    "            \"Top Attended Marker Genes\": top_genes\n",
    "        })\n",
    "        sample_idx += 1\n",
    "\n",
    "# --- Display ---\n",
    "df = pd.DataFrame(results)\n",
    "display(df.head(20))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b97b1f63-a461-4a9c-b1e9-aae96da39295",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fine-Tuning Validation: 100%|██████████████████████████████████████████████████████████████████████████████████| 367/367 [00:35<00:00, 10.43it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>True Group</th>\n",
       "      <th>Predicted Group</th>\n",
       "      <th>Top Attended Marker Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_ovarian cancer</td>\n",
       "      <td>[CD2, TRBC2, TRBC1, CD3D, CD7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>malignant cell_lung cancer</td>\n",
       "      <td>malignant cell_lung cancer</td>\n",
       "      <td>[EPCAM, KRT19, SPINT2, KRT18, MGST1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fibroblast_colorectal cancer</td>\n",
       "      <td>fibroblast_colorectal cancer</td>\n",
       "      <td>[SPARC, COL6A2, COL1A2, TPM2, COL3A1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>[CST7, CD3D, CREM, CD3E, IL32]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[TRBC2, CD3D, RGS1, IL32, PTPRC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>[TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>T cell_lung cancer</td>\n",
       "      <td>[CD2, CD7, CD3D, RGS1, IL32]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>endothelial cell_ovarian cancer</td>\n",
       "      <td>endothelial cell_ovarian cancer</td>\n",
       "      <td>[RAMP2, CLDN5, HSPG2, VWF, CRIP2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>malignant cell_uveal melanoma</td>\n",
       "      <td>malignant cell_uveal melanoma</td>\n",
       "      <td>[EFHD1, ST3GAL4, S100A1, MLANA, CITED1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>B cell_ovarian cancer</td>\n",
       "      <td>[MDK, CD24, RPL7, WFDC2, GSTP1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mononuclear phagocyte_colorectal cancer</td>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>[TYROBP, FTL, FCER1G, LYZ, AIF1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>B cell_melanoma</td>\n",
       "      <td>[TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>T cell_uveal melanoma</td>\n",
       "      <td>[CREM, CST7, CD3D, CD3E, IL32]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[CD74]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mononuclear phagocyte_uveal melanoma</td>\n",
       "      <td>mononuclear phagocyte_uveal melanoma</td>\n",
       "      <td>[MS4A7, FTL, TYROBP, CST3, CD74]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>malignant cell_ovarian cancer</td>\n",
       "      <td>[KRT8, CD24, RPL7, KRT18, KRT19]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>epithelial cell_colorectal cancer</td>\n",
       "      <td>epithelial cell_colorectal cancer</td>\n",
       "      <td>[PHGR1, S100P, LGALS4, FXYD3, KRT8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[CD79A, CD37, HLA-DRB1, MS4A1, HLA-DRA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>mononuclear phagocyte_ovarian cancer</td>\n",
       "      <td>[HLA-DPB1, TYROBP, FTL, FCER1G, HLA-DRB1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>B cell_lung cancer</td>\n",
       "      <td>[TTC19]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 True Group  \\\n",
       "0                        T cell_lung cancer   \n",
       "1                malignant cell_lung cancer   \n",
       "2              fibroblast_colorectal cancer   \n",
       "3                     T cell_uveal melanoma   \n",
       "4                        T cell_lung cancer   \n",
       "5                           B cell_melanoma   \n",
       "6                        T cell_lung cancer   \n",
       "7           endothelial cell_ovarian cancer   \n",
       "8             malignant cell_uveal melanoma   \n",
       "9             malignant cell_ovarian cancer   \n",
       "10  mononuclear phagocyte_colorectal cancer   \n",
       "11                          B cell_melanoma   \n",
       "12                    T cell_uveal melanoma   \n",
       "13                       B cell_lung cancer   \n",
       "14     mononuclear phagocyte_uveal melanoma   \n",
       "15            malignant cell_ovarian cancer   \n",
       "16        epithelial cell_colorectal cancer   \n",
       "17                       B cell_lung cancer   \n",
       "18     mononuclear phagocyte_ovarian cancer   \n",
       "19                       B cell_lung cancer   \n",
       "\n",
       "                         Predicted Group  \\\n",
       "0                  T cell_ovarian cancer   \n",
       "1             malignant cell_lung cancer   \n",
       "2           fibroblast_colorectal cancer   \n",
       "3                  T cell_uveal melanoma   \n",
       "4                     T cell_lung cancer   \n",
       "5                        B cell_melanoma   \n",
       "6                     T cell_lung cancer   \n",
       "7        endothelial cell_ovarian cancer   \n",
       "8          malignant cell_uveal melanoma   \n",
       "9                  B cell_ovarian cancer   \n",
       "10  mononuclear phagocyte_ovarian cancer   \n",
       "11                       B cell_melanoma   \n",
       "12                 T cell_uveal melanoma   \n",
       "13                    B cell_lung cancer   \n",
       "14  mononuclear phagocyte_uveal melanoma   \n",
       "15         malignant cell_ovarian cancer   \n",
       "16     epithelial cell_colorectal cancer   \n",
       "17                    B cell_lung cancer   \n",
       "18  mononuclear phagocyte_ovarian cancer   \n",
       "19                    B cell_lung cancer   \n",
       "\n",
       "                        Top Attended Marker Genes  \n",
       "0                  [CD2, TRBC2, TRBC1, CD3D, CD7]  \n",
       "1            [EPCAM, KRT19, SPINT2, KRT18, MGST1]  \n",
       "2           [SPARC, COL6A2, COL1A2, TPM2, COL3A1]  \n",
       "3                  [CST7, CD3D, CREM, CD3E, IL32]  \n",
       "4                [TRBC2, CD3D, RGS1, IL32, PTPRC]  \n",
       "5   [TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4]  \n",
       "6                    [CD2, CD7, CD3D, RGS1, IL32]  \n",
       "7               [RAMP2, CLDN5, HSPG2, VWF, CRIP2]  \n",
       "8         [EFHD1, ST3GAL4, S100A1, MLANA, CITED1]  \n",
       "9                 [MDK, CD24, RPL7, WFDC2, GSTP1]  \n",
       "10               [TYROBP, FTL, FCER1G, LYZ, AIF1]  \n",
       "11  [TMEM212, UGDH-AS1, KCNQ1OT1, METTL21A, ORC4]  \n",
       "12                 [CREM, CST7, CD3D, CD3E, IL32]  \n",
       "13                                         [CD74]  \n",
       "14               [MS4A7, FTL, TYROBP, CST3, CD74]  \n",
       "15               [KRT8, CD24, RPL7, KRT18, KRT19]  \n",
       "16            [PHGR1, S100P, LGALS4, FXYD3, KRT8]  \n",
       "17        [CD79A, CD37, HLA-DRB1, MS4A1, HLA-DRA]  \n",
       "18      [HLA-DPB1, TYROBP, FTL, FCER1G, HLA-DRB1]  \n",
       "19                                        [TTC19]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import torch\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from torch.utils.data import DataLoader, SequentialSampler\n",
    "from helical.models.scgpt.data_collator import DataCollator\n",
    "\n",
    "# Ensure eval mode\n",
    "scgpt_fine_tune.model.eval()\n",
    "vocab = scgpt_fine_tune.vocab\n",
    "id_to_gene = {v: k for k, v in vocab.items()}\n",
    "\n",
    "# --- Collator with fixed padding ---\n",
    "collator = DataCollator(\n",
    "    do_padding=True,\n",
    "    pad_token_id=vocab[\"<pad>\"],\n",
    "    pad_value=0,\n",
    "    max_length=1200,\n",
    "    do_binning=True,\n",
    "    do_mlm=False,\n",
    "    sampling=False,\n",
    "    keep_first_n_tokens=1,\n",
    ")\n",
    "\n",
    "val_loader = DataLoader(\n",
    "    validation_dataset,\n",
    "    batch_size=32,\n",
    "    sampler=SequentialSampler(validation_dataset),\n",
    "    collate_fn=collator,\n",
    "    drop_last=False,\n",
    ")\n",
    "\n",
    "# --- Prediction outputs ---\n",
    "outputs = scgpt_fine_tune.get_outputs(validation_dataset)\n",
    "y_pred = np.argmax(outputs, axis=1)\n",
    "y_true = group_labels_test.numpy() if isinstance(group_labels_test, torch.Tensor) else group_labels_test\n",
    "id_class_dict = {i: c for i, c in enumerate(label_encoder.classes_)}\n",
    "true_groups = [id_class_dict[idx] for idx in y_true]\n",
    "pred_groups = [id_class_dict[idx] for idx in y_pred]\n",
    "\n",
    "# --- Map marker genes to token IDs ---\n",
    "group_to_marker_ids = {}\n",
    "for group, genes in final_marker_dict.items():\n",
    "    marker_ids = [vocab[g] for g in genes if g in vocab]\n",
    "    group_to_marker_ids[group] = marker_ids\n",
    "\n",
    "# --- CLS attention extractor ---\n",
    "def get_cls_attention(model, input_tensor):\n",
    "    with torch.no_grad():\n",
    "        embedded = model.encoder(input_tensor)\n",
    "        attn_list = []\n",
    "        for layer in model.transformer_encoder.layers:\n",
    "            attn = layer.self_attn(embedded, embedded, embedded, need_weights=True)[1]\n",
    "            cls_attn = attn[:, 0, :]  # CLS to each token\n",
    "            attn_list.append(cls_attn)\n",
    "        attn_avg = torch.stack(attn_list).mean(dim=0)  # (B, seq)\n",
    "    return attn_avg\n",
    "\n",
    "# --- Extract top attended marker genes per sample ---\n",
    "results = []\n",
    "sample_idx = 0\n",
    "\n",
    "for batch in val_loader:\n",
    "    device = next(scgpt_fine_tune.model.parameters()).device\n",
    "    input_ids = batch[\"gene\"].to(device)\n",
    "    attn_scores = get_cls_attention(scgpt_fine_tune.model, input_ids)  # (B, seq)\n",
    "\n",
    "    for i in range(attn_scores.shape[0]):\n",
    "        sample_input = input_ids[i].cpu().numpy()\n",
    "        sample_attn = attn_scores[i].cpu().numpy()\n",
    "        true_group = true_groups[sample_idx]\n",
    "        pred_group = pred_groups[sample_idx]\n",
    "\n",
    "        marker_ids = group_to_marker_ids.get(true_group, [])\n",
    "        top_marker_attn = [\n",
    "            (id_to_gene[token_id], sample_attn[j])\n",
    "            for j, token_id in enumerate(sample_input)\n",
    "            if token_id in marker_ids\n",
    "        ]\n",
    "        top_marker_attn.sort(key=lambda x: x[1], reverse=True)\n",
    "        top_genes = [g for g, _ in top_marker_attn[:5]]\n",
    "\n",
    "        results.append({\n",
    "            \"True Group\": true_group,\n",
    "            \"Predicted Group\": pred_group,\n",
    "            \"Top Attended Marker Genes\": top_genes\n",
    "        })\n",
    "        sample_idx += 1\n",
    "\n",
    "# --- Display ---\n",
    "df = pd.DataFrame(results)\n",
    "display(df.head(20))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f13dfa94-bc37-424f-9510-43da058a6813",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
